// @wordpress/components/build/mobile/color-settings/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _native = require("@react-navigation/native");
var _element = require("@wordpress/element");
var _pickerScreen = _interopRequireDefault(require("./picker-screen"));
var _gradientPickerScreen = _interopRequireDefault(require("./gradient-picker-screen"));
var _palette = _interopRequireDefault(require("./palette.screen"));
var _bottomSheet = _interopRequireDefault(require("../bottom-sheet"));
var _bottomSheetContext = require("../bottom-sheet/bottom-sheet-context");
var _utils = require("./utils");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const ColorSettingsMemo = (0, _element.memo)(({
  defaultSettings,
  onHandleClosingBottomSheet,
  shouldEnableBottomSheetMaxHeight,
  onColorChange,
  colorValue,
  gradientValue,
  onGradientChange,
  onColorCleared,
  label,
  hideNavigation
}) => {
  (0, _element.useEffect)(() => {
    shouldEnableBottomSheetMaxHeight(true);
    onHandleClosingBottomSheet(null);
    // See https://github.com/WordPress/gutenberg/pull/41166
  }, []);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_bottomSheet.default.NavigationContainer, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_bottomSheet.default.NavigationScreen, {
      name: _utils.colorsUtils.screens.palette,
      initialParams: {
        defaultSettings,
        onColorChange,
        colorValue,
        gradientValue,
        onGradientChange,
        onColorCleared,
        label,
        hideNavigation
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_palette.default, {})
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_bottomSheet.default.NavigationScreen, {
      name: _utils.colorsUtils.screens.picker,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_pickerScreen.default, {})
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_bottomSheet.default.NavigationScreen, {
      name: _utils.colorsUtils.screens.gradientPicker,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_gradientPickerScreen.default, {})
    })]
  });
});
function ColorSettings(props) {
  const route = (0, _native.useRoute)();
  const {
    onHandleClosingBottomSheet,
    shouldEnableBottomSheetMaxHeight
  } = (0, _element.useContext)(_bottomSheetContext.BottomSheetContext);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ColorSettingsMemo, {
    onHandleClosingBottomSheet: onHandleClosingBottomSheet,
    shouldEnableBottomSheetMaxHeight: shouldEnableBottomSheetMaxHeight,
    ...props,
    ...route.params
  });
}
var _default = exports.default = ColorSettings;
