// @wordpress/block-editor/build/components/image-editor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageEditor;
var _components = require("@wordpress/components");
var _aspectRatioDropdown = _interopRequireDefault(require("./aspect-ratio-dropdown"));
var _blockControls = _interopRequireDefault(require("../block-controls"));
var _context = _interopRequireDefault(require("./context"));
var _cropper = _interopRequireDefault(require("./cropper"));
var _zoomDropdown = _interopRequireDefault(require("./zoom-dropdown"));
var _rotationButton = _interopRequireDefault(require("./rotation-button"));
var _formControls = _interopRequireDefault(require("./form-controls"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ImageEditor({
  id,
  url,
  width,
  height,
  naturalHeight,
  naturalWidth,
  onSaveImage,
  onFinishEditing,
  borderProps
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_context.default, {
    id: id,
    url: url,
    naturalWidth: naturalWidth,
    naturalHeight: naturalHeight,
    onSaveImage: onSaveImage,
    onFinishEditing: onFinishEditing,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_cropper.default, {
      borderProps: borderProps,
      url: url,
      width: width,
      height: height,
      naturalHeight: naturalHeight,
      naturalWidth: naturalWidth
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_blockControls.default, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.ToolbarGroup, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_zoomDropdown.default, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarItem, {
          children: toggleProps => /*#__PURE__*/(0, _jsxRuntime.jsx)(_aspectRatioDropdown.default, {
            toggleProps: toggleProps
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_rotationButton.default, {})]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ToolbarGroup, {
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_formControls.default, {})
      })]
    })]
  });
}
