// @wordpress/edit-post/build/components/header/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _element = require("@wordpress/element");
require("@wordpress/editor");
var _headerToolbar = _interopRequireDefault(require("./header-toolbar"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

class Header extends _element.Component {
  constructor() {
    super(...arguments);
    this.keyboardDidShow = this.keyboardDidShow.bind(this);
    this.keyboardDidHide = this.keyboardDidHide.bind(this);
    this.state = {
      isKeyboardVisible: false
    };
  }
  componentDidMount() {
    this.keyboardShowSubscription = _reactNative.Keyboard.addListener('keyboardDidShow', this.keyboardDidShow);
    this.keyboardHideSubscription = _reactNative.Keyboard.addListener('keyboardDidHide', this.keyboardDidHide);
  }
  componentWillUnmount() {
    this.keyboardShowSubscription.remove();
    this.keyboardHideSubscription.remove();
  }
  keyboardDidShow() {
    this.setState({
      isKeyboardVisible: true
    });
  }
  keyboardDidHide() {
    this.setState({
      isKeyboardVisible: false
    });
  }
  render() {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_headerToolbar.default, {
      showKeyboardHideButton: this.state.isKeyboardVisible
    });
  }
}
exports.default = Header;
