// @wordpress/dataviews/build/components/dataviews-search/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _compose = require("@wordpress/compose");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const DataViewsSearch = (0, _element.memo)(function Search({
  label
}) {
  const {
    view,
    onChangeView
  } = (0, _element.useContext)(_dataviewsContext.default);
  const [search, setSearch, debouncedSearch] = (0, _compose.useDebouncedInput)(view.search);
  (0, _element.useEffect)(() => {
    var _view$search;
    setSearch((_view$search = view.search) !== null && _view$search !== void 0 ? _view$search : '');
  }, [view.search, setSearch]);
  const onChangeViewRef = (0, _element.useRef)(onChangeView);
  const viewRef = (0, _element.useRef)(view);
  (0, _element.useEffect)(() => {
    onChangeViewRef.current = onChangeView;
    viewRef.current = view;
  }, [onChangeView, view]);
  (0, _element.useEffect)(() => {
    if (debouncedSearch !== viewRef.current?.search) {
      onChangeViewRef.current({
        ...viewRef.current,
        page: 1,
        search: debouncedSearch
      });
    }
  }, [debouncedSearch]);
  const searchLabel = label || (0, _i18n.__)('Search');
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SearchControl, {
    className: "dataviews-search",
    __nextHasNoMarginBottom: true,
    onChange: setSearch,
    value: search,
    label: searchLabel,
    placeholder: searchLabel,
    size: "compact"
  });
});
var _default = exports.default = DataViewsSearch;
