// @wordpress/block-library/build/comments-pagination-previous/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommentsPaginationPreviousEdit;
var _i18n = require("@wordpress/i18n");
var _blockEditor = require("@wordpress/block-editor");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

const arrowMap = {
  none: '',
  arrow: '←',
  chevron: '«'
};
function CommentsPaginationPreviousEdit({
  attributes: {
    label
  },
  setAttributes,
  context: {
    'comments/paginationArrow': paginationArrow
  }
}) {
  const displayArrow = arrowMap[paginationArrow];
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("a", {
    href: "#comments-pagination-previous-pseudo-link",
    onClick: event => event.preventDefault(),
    ...(0, _blockEditor.useBlockProps)(),
    children: [displayArrow && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: `wp-block-comments-pagination-previous-arrow is-arrow-${paginationArrow}`,
      children: displayArrow
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.PlainText, {
      __experimentalVersion: 2,
      tagName: "span",
      "aria-label": (0, _i18n.__)('Older comments page link'),
      placeholder: (0, _i18n.__)('Older Comments'),
      value: label,
      onChange: newLabel => setAttributes({
        label: newLabel
      })
    })]
  });
}
