// @wordpress/fields/build/fields/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "authorField", {
  enumerable: true,
  get: function () {
    return _author.default;
  }
});
Object.defineProperty(exports, "commentStatusField", {
  enumerable: true,
  get: function () {
    return _commentStatus.default;
  }
});
Object.defineProperty(exports, "dateField", {
  enumerable: true,
  get: function () {
    return _date.default;
  }
});
Object.defineProperty(exports, "featuredImageField", {
  enumerable: true,
  get: function () {
    return _featuredImage.default;
  }
});
Object.defineProperty(exports, "orderField", {
  enumerable: true,
  get: function () {
    return _order.default;
  }
});
Object.defineProperty(exports, "pageTitleField", {
  enumerable: true,
  get: function () {
    return _pageTitle.default;
  }
});
Object.defineProperty(exports, "parentField", {
  enumerable: true,
  get: function () {
    return _parent.default;
  }
});
Object.defineProperty(exports, "passwordField", {
  enumerable: true,
  get: function () {
    return _password.default;
  }
});
Object.defineProperty(exports, "patternTitleField", {
  enumerable: true,
  get: function () {
    return _patternTitle.default;
  }
});
Object.defineProperty(exports, "slugField", {
  enumerable: true,
  get: function () {
    return _slug.default;
  }
});
Object.defineProperty(exports, "statusField", {
  enumerable: true,
  get: function () {
    return _status.default;
  }
});
Object.defineProperty(exports, "templateField", {
  enumerable: true,
  get: function () {
    return _template.default;
  }
});
Object.defineProperty(exports, "templateTitleField", {
  enumerable: true,
  get: function () {
    return _templateTitle.default;
  }
});
Object.defineProperty(exports, "titleField", {
  enumerable: true,
  get: function () {
    return _title.default;
  }
});
var _slug = _interopRequireDefault(require("./slug"));
var _title = _interopRequireDefault(require("./title"));
var _pageTitle = _interopRequireDefault(require("./page-title"));
var _templateTitle = _interopRequireDefault(require("./template-title"));
var _patternTitle = _interopRequireDefault(require("./pattern-title"));
var _order = _interopRequireDefault(require("./order"));
var _featuredImage = _interopRequireDefault(require("./featured-image"));
var _template = _interopRequireDefault(require("./template"));
var _parent = _interopRequireDefault(require("./parent"));
var _password = _interopRequireDefault(require("./password"));
var _status = _interopRequireDefault(require("./status"));
var _commentStatus = _interopRequireDefault(require("./comment-status"));
var _date = _interopRequireDefault(require("./date"));
var _author = _interopRequireDefault(require("./author"));
