// @wordpress/fields/build/fields/author/author-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function AuthorView({
  item
}) {
  const {
    text,
    imageUrl
  } = (0, _data.useSelect)(select => {
    const {
      getEntityRecord
    } = select(_coreData.store);
    let user;
    if (!!item.author) {
      user = getEntityRecord('root', 'user', item.author);
    }
    return {
      imageUrl: user?.avatar_urls?.[48],
      text: user?.name
    };
  }, [item]);
  const [isImageLoaded, setIsImageLoaded] = (0, _element.useState)(false);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    alignment: "left",
    spacing: 0,
    children: [!!imageUrl && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: (0, _clsx.default)('page-templates-author-field__avatar', {
        'is-loaded': isImageLoaded
      }),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
        onLoad: () => setIsImageLoaded(true),
        alt: (0, _i18n.__)('Author avatar'),
        src: imageUrl
      })
    }), !imageUrl && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "page-templates-author-field__icon",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
        icon: _icons.commentAuthorAvatar
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: "page-templates-author-field__name",
      children: text
    })]
  });
}
var _default = exports.default = AuthorView;
