// @wordpress/editor/build/components/blog-title/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlogTitle;
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _htmlEntities = require("@wordpress/html-entities");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _blockEditor = require("@wordpress/block-editor");
var _constants = require("../../store/constants");
var _postPanelRow = _interopRequireDefault(require("../post-panel-row"));
var _store = require("../../store");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const EMPTY_OBJECT = {};
function BlogTitle() {
  const {
    editEntityRecord
  } = (0, _data.useDispatch)(_coreData.store);
  const {
    postsPageTitle,
    postsPageId,
    isTemplate,
    postSlug
  } = (0, _data.useSelect)(select => {
    const {
      getEntityRecord,
      getEditedEntityRecord,
      canUser
    } = select(_coreData.store);
    const siteSettings = canUser('read', {
      kind: 'root',
      name: 'site'
    }) ? getEntityRecord('root', 'site') : undefined;
    const _postsPageRecord = siteSettings?.page_for_posts ? getEditedEntityRecord('postType', 'page', siteSettings?.page_for_posts) : EMPTY_OBJECT;
    const {
      getEditedPostAttribute,
      getCurrentPostType
    } = select(_store.store);
    return {
      postsPageId: _postsPageRecord?.id,
      postsPageTitle: _postsPageRecord?.title,
      isTemplate: getCurrentPostType() === _constants.TEMPLATE_POST_TYPE,
      postSlug: getEditedPostAttribute('slug')
    };
  }, []);
  // Use internal state instead of a ref to make sure that the component
  // re-renders when the popover's anchor updates.
  const [popoverAnchor, setPopoverAnchor] = (0, _element.useState)(null);
  // Memoize popoverProps to avoid returning a new object every time.
  const popoverProps = (0, _element.useMemo)(() => ({
    // Anchor the popover to the middle of the entire row so that it doesn't
    // move around when the label changes.
    anchor: popoverAnchor,
    placement: 'left-start',
    offset: 36,
    shift: true
  }), [popoverAnchor]);
  if (!isTemplate || !['home', 'index'].includes(postSlug) || !postsPageId) {
    return null;
  }
  const setPostsPageTitle = newValue => {
    editEntityRecord('postType', 'page', postsPageId, {
      title: newValue
    });
  };
  const decodedTitle = (0, _htmlEntities.decodeEntities)(postsPageTitle);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_postPanelRow.default, {
    label: (0, _i18n.__)('Blog title'),
    ref: setPopoverAnchor,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Dropdown, {
      popoverProps: popoverProps,
      contentClassName: "editor-blog-title-dropdown__content",
      focusOnMount: true,
      renderToggle: ({
        isOpen,
        onToggle
      }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        size: "compact",
        variant: "tertiary",
        "aria-expanded": isOpen,
        "aria-label": (0, _i18n.sprintf)(
        // translators: %s: Current post link.
        (0, _i18n.__)('Change blog title: %s'), decodedTitle),
        onClick: onToggle,
        children: decodedTitle
      }),
      renderContent: ({
        onClose
      }) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.__experimentalInspectorPopoverHeader, {
          title: (0, _i18n.__)('Blog title'),
          onClose: onClose
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__experimentalInputControl, {
          placeholder: (0, _i18n.__)('No title'),
          size: "__unstable-large",
          value: postsPageTitle,
          onChange: (0, _compose.debounce)(setPostsPageTitle, 300),
          label: (0, _i18n.__)('Blog title'),
          help: (0, _i18n.__)('Set the Posts Page title. Appears in search results, and when the page is shared on social media.'),
          hideLabelFromVision: true
        })]
      })
    })
  });
}
