// @wordpress/block-library/build/cover/edit/cover-placeholder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CoverPlaceholder;
var _blockEditor = require("@wordpress/block-editor");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _shared = require("../shared");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function CoverPlaceholder({
  disableMediaButtons = false,
  children,
  onSelectMedia,
  onError,
  style,
  toggleUseFeaturedImage
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.MediaPlaceholder, {
    icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.BlockIcon, {
      icon: _icons.cover
    }),
    labels: {
      title: (0, _i18n.__)('Cover')
    },
    onSelect: onSelectMedia,
    accept: "image/*,video/*",
    allowedTypes: _shared.ALLOWED_MEDIA_TYPES,
    disableMediaButtons: disableMediaButtons,
    onToggleFeaturedImage: toggleUseFeaturedImage,
    onError: onError,
    style: style,
    children: children
  });
}
