// @wordpress/block-editor/build/components/block-manager/checklist.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _blockIcon = _interopRequireDefault(require("../block-icon"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BlockTypesChecklist({
  blockTypes,
  value,
  onItemChange
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
    className: "block-editor-block-manager__checklist",
    children: blockTypes.map(blockType => /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
      className: "block-editor-block-manager__checklist-item",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.CheckboxControl, {
        __nextHasNoMarginBottom: true,
        label: blockType.title,
        checked: value.includes(blockType.name),
        onChange: (...args) => onItemChange(blockType, ...args)
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockIcon.default, {
        icon: blockType.icon
      })]
    }, blockType.name))
  });
}
var _default = exports.default = BlockTypesChecklist;
