// @wordpress/edit-post/build-module/components/preferences-modal/enable-custom-fields.js
/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { Button } from '@wordpress/components';
import { useSelect } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';
import { privateApis as preferencesPrivateApis } from '@wordpress/preferences';
import { getPathAndQueryString } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { unlock } from '../../lock-unlock';
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
const {
  PreferenceBaseOption
} = unlock(preferencesPrivateApis);
function submitCustomFieldsForm() {
  const customFieldsForm = document.getElementById('toggle-custom-fields-form');

  // Ensure the referrer values is up to update with any
  customFieldsForm.querySelector('[name="_wp_http_referer"]').setAttribute('value', getPathAndQueryString(window.location.href));
  customFieldsForm.submit();
}
export function CustomFieldsConfirmation({
  willEnable
}) {
  const [isReloading, setIsReloading] = useState(false);
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx("p", {
      className: "edit-post-preferences-modal__custom-fields-confirmation-message",
      children: __('A page reload is required for this change. Make sure your content is saved before reloading.')
    }), /*#__PURE__*/_jsx(Button, {
      __next40pxDefaultSize: true,
      variant: "secondary",
      isBusy: isReloading,
      accessibleWhenDisabled: true,
      disabled: isReloading,
      onClick: () => {
        setIsReloading(true);
        submitCustomFieldsForm();
      },
      children: willEnable ? __('Show & Reload Page') : __('Hide & Reload Page')
    })]
  });
}
export default function EnableCustomFieldsOption({
  label
}) {
  const areCustomFieldsEnabled = useSelect(select => {
    return !!select(editorStore).getEditorSettings().enableCustomFields;
  }, []);
  const [isChecked, setIsChecked] = useState(areCustomFieldsEnabled);
  return /*#__PURE__*/_jsx(PreferenceBaseOption, {
    label: label,
    isChecked: isChecked,
    onChange: setIsChecked,
    children: isChecked !== areCustomFieldsEnabled && /*#__PURE__*/_jsx(CustomFieldsConfirmation, {
      willEnable: isChecked
    })
  });
}
