// @wordpress/block-library/build/column/column-preview.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _compose = require("@wordpress/compose");
var _components = require("@wordpress/components");
var _editor = _interopRequireDefault(require("./editor.scss"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ColumnPreviewItem({
  index,
  selectedColumnIndex,
  width
}) {
  const columnIndicatorStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.columnIndicator, _editor.default.columnIndicatorDark);
  const isSelectedColumn = index === selectedColumnIndex;
  const convertedWidth = (0, _components.useConvertUnitToMobile)(width);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: [isSelectedColumn && columnIndicatorStyle, {
      flex: convertedWidth
    }]
  }, index);
}
function ColumnsPreview({
  columnWidths,
  selectedColumnIndex
}) {
  const columnsPreviewStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.columnsPreview, _editor.default.columnsPreviewDark);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: columnsPreviewStyle,
    children: columnWidths.map((width, index) => {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(ColumnPreviewItem, {
        index: index,
        selectedColumnIndex: selectedColumnIndex,
        width: width
      }, index);
    })
  });
}
var _default = exports.default = ColumnsPreview;
