// @wordpress/block-library/build/comments/edit/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommentsEdit;
var _blockEditor = require("@wordpress/block-editor");
var _commentsInspectorControls = _interopRequireDefault(require("./comments-inspector-controls"));
var _commentsLegacy = _interopRequireDefault(require("./comments-legacy"));
var _template = _interopRequireDefault(require("./template"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function CommentsEdit(props) {
  const {
    attributes,
    setAttributes
  } = props;
  const {
    tagName: TagName,
    legacy
  } = attributes;
  const blockProps = (0, _blockEditor.useBlockProps)();
  const innerBlocksProps = (0, _blockEditor.useInnerBlocksProps)(blockProps, {
    template: _template.default
  });
  if (legacy) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_commentsLegacy.default, {
      ...props
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_commentsInspectorControls.default, {
      attributes: attributes,
      setAttributes: setAttributes
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(TagName, {
      ...innerBlocksProps
    })]
  });
}
