// @wordpress/edit-post/build/components/back-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _editor = require("@wordpress/editor");
var _components = require("@wordpress/components");
var _fullscreenModeClose = _interopRequireDefault(require("./fullscreen-mode-close"));
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  BackButton: BackButtonFill
} = (0, _lockUnlock.unlock)(_editor.privateApis);
const slideX = {
  hidden: {
    x: '-100%'
  },
  distractionFreeInactive: {
    x: 0
  },
  hover: {
    x: 0,
    transition: {
      type: 'tween',
      delay: 0.2
    }
  }
};
function BackButton({
  initialPost
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(BackButtonFill, {
    children: ({
      length
    }) => length <= 1 && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.__unstableMotion.div, {
      variants: slideX,
      transition: {
        type: 'tween',
        delay: 0.8
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fullscreenModeClose.default, {
        showTooltip: true,
        initialPost: initialPost
      })
    })
  });
}
var _default = exports.default = BackButton;
