// @wordpress/editor/build/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  VisualEditorGlobalKeyboardShortcuts: true,
  TextEditorGlobalKeyboardShortcuts: true,
  EditorKeyboardShortcuts: true,
  AutosaveMonitor: true,
  DocumentBar: true,
  DocumentOutline: true,
  DocumentOutlineCheck: true,
  EditorKeyboardShortcutsRegister: true,
  EditorHistoryRedo: true,
  EditorHistoryUndo: true,
  EditorNotices: true,
  EditorSnackbars: true,
  EntitiesSavedStates: true,
  useEntitiesSavedStatesIsDirty: true,
  ErrorBoundary: true,
  LocalAutosaveMonitor: true,
  PageAttributesCheck: true,
  PageAttributesOrder: true,
  PageAttributesPanel: true,
  PageAttributesParent: true,
  PageTemplate: true,
  PluginDocumentSettingPanel: true,
  PluginBlockSettingsMenuItem: true,
  PluginMoreMenuItem: true,
  PluginPostPublishPanel: true,
  PluginPostStatusInfo: true,
  PluginPrePublishPanel: true,
  PluginPreviewMenuItem: true,
  PluginSidebar: true,
  PluginSidebarMoreMenuItem: true,
  PostTemplatePanel: true,
  PostAuthor: true,
  PostAuthorCheck: true,
  PostAuthorPanel: true,
  PostComments: true,
  PostDiscussionPanel: true,
  PostExcerpt: true,
  PostExcerptCheck: true,
  PostExcerptPanel: true,
  PostFeaturedImage: true,
  PostFeaturedImageCheck: true,
  PostFeaturedImagePanel: true,
  PostFormat: true,
  PostFormatCheck: true,
  PostLastRevision: true,
  PostLastRevisionCheck: true,
  PostLastRevisionPanel: true,
  PostLockedModal: true,
  PostPendingStatus: true,
  PostPendingStatusCheck: true,
  PostPingbacks: true,
  PostPreviewButton: true,
  PostPublishButton: true,
  PostPublishButtonLabel: true,
  PostPublishPanel: true,
  PostSavedState: true,
  PostSchedule: true,
  PostScheduleCheck: true,
  PostScheduleLabel: true,
  usePostScheduleLabel: true,
  PostSchedulePanel: true,
  PostSticky: true,
  PostStickyCheck: true,
  PostSwitchToDraftButton: true,
  PostSyncStatus: true,
  PostTaxonomies: true,
  PostTaxonomiesFlatTermSelector: true,
  PostTaxonomiesHierarchicalTermSelector: true,
  PostTaxonomiesCheck: true,
  PostTaxonomiesPanel: true,
  PostTextEditor: true,
  PostTitle: true,
  PostTitleRaw: true,
  PostTrash: true,
  PostTrashCheck: true,
  PostTypeSupportCheck: true,
  PostURL: true,
  PostURLCheck: true,
  PostURLLabel: true,
  usePostURLLabel: true,
  PostURLPanel: true,
  PostVisibility: true,
  PostVisibilityLabel: true,
  usePostVisibilityLabel: true,
  PostVisibilityCheck: true,
  TableOfContents: true,
  ThemeSupportCheck: true,
  UnsavedChangesWarning: true,
  WordCount: true,
  TimeToRead: true,
  CharacterCount: true,
  EditorProvider: true
};
Object.defineProperty(exports, "AutosaveMonitor", {
  enumerable: true,
  get: function () {
    return _autosaveMonitor.default;
  }
});
Object.defineProperty(exports, "CharacterCount", {
  enumerable: true,
  get: function () {
    return _characterCount.default;
  }
});
Object.defineProperty(exports, "DocumentBar", {
  enumerable: true,
  get: function () {
    return _documentBar.default;
  }
});
Object.defineProperty(exports, "DocumentOutline", {
  enumerable: true,
  get: function () {
    return _documentOutline.default;
  }
});
Object.defineProperty(exports, "DocumentOutlineCheck", {
  enumerable: true,
  get: function () {
    return _check.default;
  }
});
Object.defineProperty(exports, "EditorHistoryRedo", {
  enumerable: true,
  get: function () {
    return _redo.default;
  }
});
Object.defineProperty(exports, "EditorHistoryUndo", {
  enumerable: true,
  get: function () {
    return _undo.default;
  }
});
Object.defineProperty(exports, "EditorKeyboardShortcuts", {
  enumerable: true,
  get: function () {
    return _globalKeyboardShortcuts.default;
  }
});
Object.defineProperty(exports, "EditorKeyboardShortcutsRegister", {
  enumerable: true,
  get: function () {
    return _registerShortcuts.default;
  }
});
Object.defineProperty(exports, "EditorNotices", {
  enumerable: true,
  get: function () {
    return _editorNotices.default;
  }
});
Object.defineProperty(exports, "EditorProvider", {
  enumerable: true,
  get: function () {
    return _provider.default;
  }
});
Object.defineProperty(exports, "EditorSnackbars", {
  enumerable: true,
  get: function () {
    return _editorSnackbars.default;
  }
});
Object.defineProperty(exports, "EntitiesSavedStates", {
  enumerable: true,
  get: function () {
    return _entitiesSavedStates.default;
  }
});
Object.defineProperty(exports, "ErrorBoundary", {
  enumerable: true,
  get: function () {
    return _errorBoundary.default;
  }
});
Object.defineProperty(exports, "LocalAutosaveMonitor", {
  enumerable: true,
  get: function () {
    return _localAutosaveMonitor.default;
  }
});
Object.defineProperty(exports, "PageAttributesCheck", {
  enumerable: true,
  get: function () {
    return _check2.default;
  }
});
Object.defineProperty(exports, "PageAttributesOrder", {
  enumerable: true,
  get: function () {
    return _order.default;
  }
});
Object.defineProperty(exports, "PageAttributesPanel", {
  enumerable: true,
  get: function () {
    return _panel.default;
  }
});
Object.defineProperty(exports, "PageAttributesParent", {
  enumerable: true,
  get: function () {
    return _parent.default;
  }
});
Object.defineProperty(exports, "PageTemplate", {
  enumerable: true,
  get: function () {
    return _classicTheme.default;
  }
});
Object.defineProperty(exports, "PluginBlockSettingsMenuItem", {
  enumerable: true,
  get: function () {
    return _pluginBlockSettingsMenuItem.default;
  }
});
Object.defineProperty(exports, "PluginDocumentSettingPanel", {
  enumerable: true,
  get: function () {
    return _pluginDocumentSettingPanel.default;
  }
});
Object.defineProperty(exports, "PluginMoreMenuItem", {
  enumerable: true,
  get: function () {
    return _pluginMoreMenuItem.default;
  }
});
Object.defineProperty(exports, "PluginPostPublishPanel", {
  enumerable: true,
  get: function () {
    return _pluginPostPublishPanel.default;
  }
});
Object.defineProperty(exports, "PluginPostStatusInfo", {
  enumerable: true,
  get: function () {
    return _pluginPostStatusInfo.default;
  }
});
Object.defineProperty(exports, "PluginPrePublishPanel", {
  enumerable: true,
  get: function () {
    return _pluginPrePublishPanel.default;
  }
});
Object.defineProperty(exports, "PluginPreviewMenuItem", {
  enumerable: true,
  get: function () {
    return _pluginPreviewMenuItem.default;
  }
});
Object.defineProperty(exports, "PluginSidebar", {
  enumerable: true,
  get: function () {
    return _pluginSidebar.default;
  }
});
Object.defineProperty(exports, "PluginSidebarMoreMenuItem", {
  enumerable: true,
  get: function () {
    return _pluginSidebarMoreMenuItem.default;
  }
});
Object.defineProperty(exports, "PostAuthor", {
  enumerable: true,
  get: function () {
    return _postAuthor.default;
  }
});
Object.defineProperty(exports, "PostAuthorCheck", {
  enumerable: true,
  get: function () {
    return _check3.default;
  }
});
Object.defineProperty(exports, "PostAuthorPanel", {
  enumerable: true,
  get: function () {
    return _panel3.default;
  }
});
Object.defineProperty(exports, "PostComments", {
  enumerable: true,
  get: function () {
    return _postComments.default;
  }
});
Object.defineProperty(exports, "PostDiscussionPanel", {
  enumerable: true,
  get: function () {
    return _panel4.default;
  }
});
Object.defineProperty(exports, "PostExcerpt", {
  enumerable: true,
  get: function () {
    return _postExcerpt.default;
  }
});
Object.defineProperty(exports, "PostExcerptCheck", {
  enumerable: true,
  get: function () {
    return _check4.default;
  }
});
Object.defineProperty(exports, "PostExcerptPanel", {
  enumerable: true,
  get: function () {
    return _panel5.default;
  }
});
Object.defineProperty(exports, "PostFeaturedImage", {
  enumerable: true,
  get: function () {
    return _postFeaturedImage.default;
  }
});
Object.defineProperty(exports, "PostFeaturedImageCheck", {
  enumerable: true,
  get: function () {
    return _check5.default;
  }
});
Object.defineProperty(exports, "PostFeaturedImagePanel", {
  enumerable: true,
  get: function () {
    return _panel6.default;
  }
});
Object.defineProperty(exports, "PostFormat", {
  enumerable: true,
  get: function () {
    return _postFormat.default;
  }
});
Object.defineProperty(exports, "PostFormatCheck", {
  enumerable: true,
  get: function () {
    return _check6.default;
  }
});
Object.defineProperty(exports, "PostLastRevision", {
  enumerable: true,
  get: function () {
    return _postLastRevision.default;
  }
});
Object.defineProperty(exports, "PostLastRevisionCheck", {
  enumerable: true,
  get: function () {
    return _check7.default;
  }
});
Object.defineProperty(exports, "PostLastRevisionPanel", {
  enumerable: true,
  get: function () {
    return _panel7.default;
  }
});
Object.defineProperty(exports, "PostLockedModal", {
  enumerable: true,
  get: function () {
    return _postLockedModal.default;
  }
});
Object.defineProperty(exports, "PostPendingStatus", {
  enumerable: true,
  get: function () {
    return _postPendingStatus.default;
  }
});
Object.defineProperty(exports, "PostPendingStatusCheck", {
  enumerable: true,
  get: function () {
    return _check8.default;
  }
});
Object.defineProperty(exports, "PostPingbacks", {
  enumerable: true,
  get: function () {
    return _postPingbacks.default;
  }
});
Object.defineProperty(exports, "PostPreviewButton", {
  enumerable: true,
  get: function () {
    return _postPreviewButton.default;
  }
});
Object.defineProperty(exports, "PostPublishButton", {
  enumerable: true,
  get: function () {
    return _postPublishButton.default;
  }
});
Object.defineProperty(exports, "PostPublishButtonLabel", {
  enumerable: true,
  get: function () {
    return _label.default;
  }
});
Object.defineProperty(exports, "PostPublishPanel", {
  enumerable: true,
  get: function () {
    return _postPublishPanel.default;
  }
});
Object.defineProperty(exports, "PostSavedState", {
  enumerable: true,
  get: function () {
    return _postSavedState.default;
  }
});
Object.defineProperty(exports, "PostSchedule", {
  enumerable: true,
  get: function () {
    return _postSchedule.default;
  }
});
Object.defineProperty(exports, "PostScheduleCheck", {
  enumerable: true,
  get: function () {
    return _check9.default;
  }
});
Object.defineProperty(exports, "PostScheduleLabel", {
  enumerable: true,
  get: function () {
    return _label2.default;
  }
});
Object.defineProperty(exports, "PostSchedulePanel", {
  enumerable: true,
  get: function () {
    return _panel8.default;
  }
});
Object.defineProperty(exports, "PostSticky", {
  enumerable: true,
  get: function () {
    return _postSticky.default;
  }
});
Object.defineProperty(exports, "PostStickyCheck", {
  enumerable: true,
  get: function () {
    return _check10.default;
  }
});
Object.defineProperty(exports, "PostSwitchToDraftButton", {
  enumerable: true,
  get: function () {
    return _postSwitchToDraftButton.default;
  }
});
Object.defineProperty(exports, "PostSyncStatus", {
  enumerable: true,
  get: function () {
    return _postSyncStatus.default;
  }
});
Object.defineProperty(exports, "PostTaxonomies", {
  enumerable: true,
  get: function () {
    return _postTaxonomies.default;
  }
});
Object.defineProperty(exports, "PostTaxonomiesCheck", {
  enumerable: true,
  get: function () {
    return _check11.default;
  }
});
Object.defineProperty(exports, "PostTaxonomiesFlatTermSelector", {
  enumerable: true,
  get: function () {
    return _flatTermSelector.FlatTermSelector;
  }
});
Object.defineProperty(exports, "PostTaxonomiesHierarchicalTermSelector", {
  enumerable: true,
  get: function () {
    return _hierarchicalTermSelector.HierarchicalTermSelector;
  }
});
Object.defineProperty(exports, "PostTaxonomiesPanel", {
  enumerable: true,
  get: function () {
    return _panel9.default;
  }
});
Object.defineProperty(exports, "PostTemplatePanel", {
  enumerable: true,
  get: function () {
    return _panel2.default;
  }
});
Object.defineProperty(exports, "PostTextEditor", {
  enumerable: true,
  get: function () {
    return _postTextEditor.default;
  }
});
Object.defineProperty(exports, "PostTitle", {
  enumerable: true,
  get: function () {
    return _postTitle.default;
  }
});
Object.defineProperty(exports, "PostTitleRaw", {
  enumerable: true,
  get: function () {
    return _postTitleRaw.default;
  }
});
Object.defineProperty(exports, "PostTrash", {
  enumerable: true,
  get: function () {
    return _postTrash.default;
  }
});
Object.defineProperty(exports, "PostTrashCheck", {
  enumerable: true,
  get: function () {
    return _check12.default;
  }
});
Object.defineProperty(exports, "PostTypeSupportCheck", {
  enumerable: true,
  get: function () {
    return _postTypeSupportCheck.default;
  }
});
Object.defineProperty(exports, "PostURL", {
  enumerable: true,
  get: function () {
    return _postUrl.default;
  }
});
Object.defineProperty(exports, "PostURLCheck", {
  enumerable: true,
  get: function () {
    return _check13.default;
  }
});
Object.defineProperty(exports, "PostURLLabel", {
  enumerable: true,
  get: function () {
    return _label3.default;
  }
});
Object.defineProperty(exports, "PostURLPanel", {
  enumerable: true,
  get: function () {
    return _panel10.default;
  }
});
Object.defineProperty(exports, "PostVisibility", {
  enumerable: true,
  get: function () {
    return _postVisibility.default;
  }
});
Object.defineProperty(exports, "PostVisibilityCheck", {
  enumerable: true,
  get: function () {
    return _check14.default;
  }
});
Object.defineProperty(exports, "PostVisibilityLabel", {
  enumerable: true,
  get: function () {
    return _label4.default;
  }
});
Object.defineProperty(exports, "TableOfContents", {
  enumerable: true,
  get: function () {
    return _tableOfContents.default;
  }
});
exports.TextEditorGlobalKeyboardShortcuts = void 0;
Object.defineProperty(exports, "ThemeSupportCheck", {
  enumerable: true,
  get: function () {
    return _themeSupportCheck.default;
  }
});
Object.defineProperty(exports, "TimeToRead", {
  enumerable: true,
  get: function () {
    return _timeToRead.default;
  }
});
Object.defineProperty(exports, "UnsavedChangesWarning", {
  enumerable: true,
  get: function () {
    return _unsavedChangesWarning.default;
  }
});
exports.VisualEditorGlobalKeyboardShortcuts = void 0;
Object.defineProperty(exports, "WordCount", {
  enumerable: true,
  get: function () {
    return _wordCount.default;
  }
});
Object.defineProperty(exports, "useEntitiesSavedStatesIsDirty", {
  enumerable: true,
  get: function () {
    return _useIsDirty.useIsDirty;
  }
});
Object.defineProperty(exports, "usePostScheduleLabel", {
  enumerable: true,
  get: function () {
    return _label2.usePostScheduleLabel;
  }
});
Object.defineProperty(exports, "usePostURLLabel", {
  enumerable: true,
  get: function () {
    return _label3.usePostURLLabel;
  }
});
Object.defineProperty(exports, "usePostVisibilityLabel", {
  enumerable: true,
  get: function () {
    return _label4.usePostVisibilityLabel;
  }
});
var _globalKeyboardShortcuts = _interopRequireDefault(require("./global-keyboard-shortcuts"));
var _autocompleters = require("./autocompleters");
Object.keys(_autocompleters).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _autocompleters[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _autocompleters[key];
    }
  });
});
var _autosaveMonitor = _interopRequireDefault(require("./autosave-monitor"));
var _documentBar = _interopRequireDefault(require("./document-bar"));
var _documentOutline = _interopRequireDefault(require("./document-outline"));
var _check = _interopRequireDefault(require("./document-outline/check"));
var _registerShortcuts = _interopRequireDefault(require("./global-keyboard-shortcuts/register-shortcuts"));
var _redo = _interopRequireDefault(require("./editor-history/redo"));
var _undo = _interopRequireDefault(require("./editor-history/undo"));
var _editorNotices = _interopRequireDefault(require("./editor-notices"));
var _editorSnackbars = _interopRequireDefault(require("./editor-snackbars"));
var _entitiesSavedStates = _interopRequireDefault(require("./entities-saved-states"));
var _useIsDirty = require("./entities-saved-states/hooks/use-is-dirty");
var _errorBoundary = _interopRequireDefault(require("./error-boundary"));
var _localAutosaveMonitor = _interopRequireDefault(require("./local-autosave-monitor"));
var _check2 = _interopRequireDefault(require("./page-attributes/check"));
var _order = _interopRequireDefault(require("./page-attributes/order"));
var _panel = _interopRequireDefault(require("./page-attributes/panel"));
var _parent = _interopRequireDefault(require("./page-attributes/parent"));
var _classicTheme = _interopRequireDefault(require("./post-template/classic-theme"));
var _pluginDocumentSettingPanel = _interopRequireDefault(require("./plugin-document-setting-panel"));
var _pluginBlockSettingsMenuItem = _interopRequireDefault(require("./block-settings-menu/plugin-block-settings-menu-item"));
var _pluginMoreMenuItem = _interopRequireDefault(require("./plugin-more-menu-item"));
var _pluginPostPublishPanel = _interopRequireDefault(require("./plugin-post-publish-panel"));
var _pluginPostStatusInfo = _interopRequireDefault(require("./plugin-post-status-info"));
var _pluginPrePublishPanel = _interopRequireDefault(require("./plugin-pre-publish-panel"));
var _pluginPreviewMenuItem = _interopRequireDefault(require("./plugin-preview-menu-item"));
var _pluginSidebar = _interopRequireDefault(require("./plugin-sidebar"));
var _pluginSidebarMoreMenuItem = _interopRequireDefault(require("./plugin-sidebar-more-menu-item"));
var _panel2 = _interopRequireDefault(require("./post-template/panel"));
var _postAuthor = _interopRequireDefault(require("./post-author"));
var _check3 = _interopRequireDefault(require("./post-author/check"));
var _panel3 = _interopRequireDefault(require("./post-author/panel"));
var _postComments = _interopRequireDefault(require("./post-comments"));
var _panel4 = _interopRequireDefault(require("./post-discussion/panel"));
var _postExcerpt = _interopRequireDefault(require("./post-excerpt"));
var _check4 = _interopRequireDefault(require("./post-excerpt/check"));
var _panel5 = _interopRequireDefault(require("./post-excerpt/panel"));
var _postFeaturedImage = _interopRequireDefault(require("./post-featured-image"));
var _check5 = _interopRequireDefault(require("./post-featured-image/check"));
var _panel6 = _interopRequireDefault(require("./post-featured-image/panel"));
var _postFormat = _interopRequireDefault(require("./post-format"));
var _check6 = _interopRequireDefault(require("./post-format/check"));
var _postLastRevision = _interopRequireDefault(require("./post-last-revision"));
var _check7 = _interopRequireDefault(require("./post-last-revision/check"));
var _panel7 = _interopRequireDefault(require("./post-last-revision/panel"));
var _postLockedModal = _interopRequireDefault(require("./post-locked-modal"));
var _postPendingStatus = _interopRequireDefault(require("./post-pending-status"));
var _check8 = _interopRequireDefault(require("./post-pending-status/check"));
var _postPingbacks = _interopRequireDefault(require("./post-pingbacks"));
var _postPreviewButton = _interopRequireDefault(require("./post-preview-button"));
var _postPublishButton = _interopRequireDefault(require("./post-publish-button"));
var _label = _interopRequireDefault(require("./post-publish-button/label"));
var _postPublishPanel = _interopRequireDefault(require("./post-publish-panel"));
var _postSavedState = _interopRequireDefault(require("./post-saved-state"));
var _postSchedule = _interopRequireDefault(require("./post-schedule"));
var _check9 = _interopRequireDefault(require("./post-schedule/check"));
var _label2 = _interopRequireWildcard(require("./post-schedule/label"));
var _panel8 = _interopRequireDefault(require("./post-schedule/panel"));
var _postSticky = _interopRequireDefault(require("./post-sticky"));
var _check10 = _interopRequireDefault(require("./post-sticky/check"));
var _postSwitchToDraftButton = _interopRequireDefault(require("./post-switch-to-draft-button"));
var _postSyncStatus = _interopRequireDefault(require("./post-sync-status"));
var _postTaxonomies = _interopRequireDefault(require("./post-taxonomies"));
var _flatTermSelector = require("./post-taxonomies/flat-term-selector");
var _hierarchicalTermSelector = require("./post-taxonomies/hierarchical-term-selector");
var _check11 = _interopRequireDefault(require("./post-taxonomies/check"));
var _panel9 = _interopRequireDefault(require("./post-taxonomies/panel"));
var _postTextEditor = _interopRequireDefault(require("./post-text-editor"));
var _postTitle = _interopRequireDefault(require("./post-title"));
var _postTitleRaw = _interopRequireDefault(require("./post-title/post-title-raw"));
var _postTrash = _interopRequireDefault(require("./post-trash"));
var _check12 = _interopRequireDefault(require("./post-trash/check"));
var _postTypeSupportCheck = _interopRequireDefault(require("./post-type-support-check"));
var _postUrl = _interopRequireDefault(require("./post-url"));
var _check13 = _interopRequireDefault(require("./post-url/check"));
var _label3 = _interopRequireWildcard(require("./post-url/label"));
var _panel10 = _interopRequireDefault(require("./post-url/panel"));
var _postVisibility = _interopRequireDefault(require("./post-visibility"));
var _label4 = _interopRequireWildcard(require("./post-visibility/label"));
var _check14 = _interopRequireDefault(require("./post-visibility/check"));
var _tableOfContents = _interopRequireDefault(require("./table-of-contents"));
var _themeSupportCheck = _interopRequireDefault(require("./theme-support-check"));
var _unsavedChangesWarning = _interopRequireDefault(require("./unsaved-changes-warning"));
var _wordCount = _interopRequireDefault(require("./word-count"));
var _timeToRead = _interopRequireDefault(require("./time-to-read"));
var _characterCount = _interopRequireDefault(require("./character-count"));
var _provider = _interopRequireDefault(require("./provider"));
var _deprecated = require("./deprecated");
Object.keys(_deprecated).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _deprecated[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _deprecated[key];
    }
  });
});
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * Internal dependencies
 */

// Block Creation Components.

// Post Related Components.

// State Related Components.

/**
 * Handles the keyboard shortcuts for the editor.
 *
 * It provides functionality for various keyboard shortcuts such as toggling editor mode,
 * toggling distraction-free mode, undo/redo, saving the post, toggling list view,
 * and toggling the sidebar.
 */
const VisualEditorGlobalKeyboardShortcuts = exports.VisualEditorGlobalKeyboardShortcuts = _globalKeyboardShortcuts.default;

/**
 * Handles the keyboard shortcuts for the editor.
 *
 * It provides functionality for various keyboard shortcuts such as toggling editor mode,
 * toggling distraction-free mode, undo/redo, saving the post, toggling list view,
 * and toggling the sidebar.
 */
const TextEditorGlobalKeyboardShortcuts = exports.TextEditorGlobalKeyboardShortcuts = _globalKeyboardShortcuts.default;
