// @wordpress/block-editor/build/components/warning/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _components = require("@wordpress/components");
var _compose = require("@wordpress/compose");
var _blocks = require("@wordpress/blocks");
var _style = _interopRequireDefault(require("./style.scss"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function Warning({
  actions,
  title,
  message,
  icon,
  iconClass,
  preferredColorScheme,
  getStylesFromColorScheme,
  containerStyle: extraContainerStyle,
  titleStyle: extraTitleStyle,
  messageStyle: extraMessageStyle,
  ...viewProps
}) {
  icon = icon && (0, _blocks.normalizeIconObject)(icon);
  const internalIconClass = 'warning-icon' + '-' + preferredColorScheme;
  const containerStyle = [getStylesFromColorScheme(_style.default.container, _style.default.containerDark), extraContainerStyle];
  const titleStyle = [getStylesFromColorScheme(_style.default.title, _style.default.titleDark), extraTitleStyle];
  const messageStyle = [getStylesFromColorScheme(_style.default.message, _style.default.messageDark), extraMessageStyle];
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: containerStyle,
    ...viewProps,
    children: [icon && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
      style: _style.default.icon,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
        className: iconClass || internalIconClass,
        icon: icon && icon.src ? icon.src : icon
      })
    }), title && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: titleStyle,
      children: title
    }), message && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: messageStyle,
      children: message
    }), actions]
  });
}
var _default = exports.default = (0, _compose.withPreferredColorScheme)(Warning);
