// @wordpress/block-library/build/query/variations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _icons = require("./icons");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const variations = [{
  name: 'title-date',
  title: (0, _i18n.__)('Title & Date'),
  icon: _icons.titleDate,
  attributes: {},
  innerBlocks: [['core/post-template', {}, [['core/post-title'], ['core/post-date']]], ['core/query-pagination'], ['core/query-no-results']],
  scope: ['block']
}, {
  name: 'title-excerpt',
  title: (0, _i18n.__)('Title & Excerpt'),
  icon: _icons.titleExcerpt,
  attributes: {},
  innerBlocks: [['core/post-template', {}, [['core/post-title'], ['core/post-excerpt']]], ['core/query-pagination'], ['core/query-no-results']],
  scope: ['block']
}, {
  name: 'title-date-excerpt',
  title: (0, _i18n.__)('Title, Date, & Excerpt'),
  icon: _icons.titleDateExcerpt,
  attributes: {},
  innerBlocks: [['core/post-template', {}, [['core/post-title'], ['core/post-date'], ['core/post-excerpt']]], ['core/query-pagination'], ['core/query-no-results']],
  scope: ['block']
}, {
  name: 'image-date-title',
  title: (0, _i18n.__)('Image, Date, & Title'),
  icon: _icons.imageDateTitle,
  attributes: {},
  innerBlocks: [['core/post-template', {}, [['core/post-featured-image'], ['core/post-date'], ['core/post-title']]], ['core/query-pagination'], ['core/query-no-results']],
  scope: ['block']
}];
var _default = exports.default = variations;
