// @wordpress/plugins/build/components/plugin-area/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _memize = _interopRequireDefault(require("memize"));
var _element = require("@wordpress/element");
var _hooks = require("@wordpress/hooks");
var _isShallowEqual = _interopRequireDefault(require("@wordpress/is-shallow-equal"));
var _pluginContext = require("../plugin-context");
var _pluginErrorBoundary = require("../plugin-error-boundary");
var _api = require("../../api");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const getPluginContext = (0, _memize.default)((icon, name) => ({
  icon,
  name
}));

/**
 * A component that renders all plugin fills in a hidden div.
 *
 * @param  props
 * @param  props.scope
 * @param  props.onError
 * @example
 * ```js
 * // Using ES5 syntax
 * var el = React.createElement;
 * var PluginArea = wp.plugins.PluginArea;
 *
 * function Layout() {
 * 	return el(
 * 		'div',
 * 		{ scope: 'my-page' },
 * 		'Content of the page',
 * 		PluginArea
 * 	);
 * }
 * ```
 *
 * @example
 * ```js
 * // Using ESNext syntax
 * import { PluginArea } from '@wordpress/plugins';
 *
 * const Layout = () => (
 * 	<div>
 * 		Content of the page
 * 		<PluginArea scope="my-page" />
 * 	</div>
 * );
 * ```
 *
 * @return {Component} The component to be rendered.
 */
function PluginArea({
  scope,
  onError
}) {
  const store = (0, _element.useMemo)(() => {
    let lastValue = [];
    return {
      subscribe(listener) {
        (0, _hooks.addAction)('plugins.pluginRegistered', 'core/plugins/plugin-area/plugins-registered', listener);
        (0, _hooks.addAction)('plugins.pluginUnregistered', 'core/plugins/plugin-area/plugins-unregistered', listener);
        return () => {
          (0, _hooks.removeAction)('plugins.pluginRegistered', 'core/plugins/plugin-area/plugins-registered');
          (0, _hooks.removeAction)('plugins.pluginUnregistered', 'core/plugins/plugin-area/plugins-unregistered');
        };
      },
      getValue() {
        const nextValue = (0, _api.getPlugins)(scope);
        if (!(0, _isShallowEqual.default)(lastValue, nextValue)) {
          lastValue = nextValue;
        }
        return lastValue;
      }
    };
  }, [scope]);
  const plugins = (0, _element.useSyncExternalStore)(store.subscribe, store.getValue, store.getValue);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      display: 'none'
    },
    children: plugins.map(({
      icon,
      name,
      render: Plugin
    }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_pluginContext.PluginContextProvider, {
      value: getPluginContext(icon, name),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_pluginErrorBoundary.PluginErrorBoundary, {
        name: name,
        onError: onError,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Plugin, {})
      })
    }, name))
  });
}
var _default = exports.default = PluginArea;
