// @wordpress/components/build/mobile/bottom-sheet/button.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _styles = _interopRequireDefault(require("./styles.scss"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

const BottomSheetButton = ({
  onPress,
  disabled,
  text,
  color
}) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.TouchableOpacity, {
  accessible: true,
  onPress: onPress,
  disabled: disabled,
  children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: {
      flexDirection: 'row',
      justifyContent: 'center'
    },
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: {
        ..._styles.default.buttonText,
        color
      },
      children: text
    })
  })
});
var _default = exports.default = BottomSheetButton;
