// @wordpress/components/build/mobile/bottom-sheet/radio-cell.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BottomSheetRadioCell;
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _compose = require("@wordpress/compose");
var _cell = _interopRequireDefault(require("./cell"));
var _styles = _interopRequireDefault(require("./styles.scss"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BottomSheetRadioCell(props) {
  const {
    selected,
    ...cellProps
  } = props;
  const selectedIconStyle = (0, _compose.usePreferredColorSchemeStyle)(_styles.default.selectedIcon, _styles.default.selectedIconDark);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_cell.default, {
    ...cellProps,
    accessibilityRole: "radio",
    accessibilityState: {
      selected
    },
    accessibilityHint: /* translators: accessibility text (hint for selecting option) */
    (0, _i18n.__)('Double tap to select the option'),
    editable: false,
    value: "",
    showLockIcon: selected,
    children: selected && /*#__PURE__*/(0, _jsxRuntime.jsx)(_icons.Icon, {
      icon: _icons.check,
      style: selectedIconStyle
    })
  });
}
