// @wordpress/fields/build/fields/title/view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseTitleView = BaseTitleView;
exports.default = TitleView;
var _clsx = _interopRequireDefault(require("clsx"));
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _utils = require("../../actions/utils");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BaseTitleView({
  item,
  className,
  children
}) {
  const renderedTitle = (0, _utils.getItemTitle)(item);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    className: (0, _clsx.default)('fields-field__title', className),
    alignment: "center",
    justify: "flex-start",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      children: renderedTitle || (0, _i18n.__)('(no title)')
    }), children]
  });
}
function TitleView({
  item
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(BaseTitleView, {
    item: item
  });
}
