// @wordpress/block-library/build/video/save.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = save;
var _blockEditor = require("@wordpress/block-editor");
var _tracks = _interopRequireDefault(require("./tracks"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function save({
  attributes
}) {
  const {
    autoplay,
    caption,
    controls,
    loop,
    muted,
    poster,
    preload,
    src,
    playsInline,
    tracks
  } = attributes;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("figure", {
    ..._blockEditor.useBlockProps.save(),
    children: [src && /*#__PURE__*/(0, _jsxRuntime.jsx)("video", {
      autoPlay: autoplay,
      controls: controls,
      loop: loop,
      muted: muted,
      poster: poster,
      preload: preload !== 'metadata' ? preload : undefined,
      src: src,
      playsInline: playsInline,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_tracks.default, {
        tracks: tracks
      })
    }), !_blockEditor.RichText.isEmpty(caption) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.RichText.Content, {
      className: (0, _blockEditor.__experimentalGetElementClassName)('caption'),
      tagName: "figcaption",
      value: caption
    })]
  });
}
