// @wordpress/block-editor/build/components/inserter/block-patterns-tab/patterns-filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PatternsFilter = PatternsFilter;
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _element = require("@wordpress/element");
var _utils = require("./utils");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const getShouldDisableSyncFilter = sourceFilter => sourceFilter !== 'all' && sourceFilter !== 'user';
const getShouldHideSourcesFilter = category => {
  return category.name === _utils.myPatternsCategory.name;
};
const PATTERN_SOURCE_MENU_OPTIONS = [{
  value: 'all',
  label: (0, _i18n._x)('All', 'patterns')
}, {
  value: _utils.INSERTER_PATTERN_TYPES.directory,
  label: (0, _i18n.__)('Pattern Directory')
}, {
  value: _utils.INSERTER_PATTERN_TYPES.theme,
  label: (0, _i18n.__)('Theme & Plugins')
}, {
  value: _utils.INSERTER_PATTERN_TYPES.user,
  label: (0, _i18n.__)('User')
}];
function PatternsFilter({
  setPatternSyncFilter,
  setPatternSourceFilter,
  patternSyncFilter,
  patternSourceFilter,
  scrollContainerRef,
  category
}) {
  // If the category is `myPatterns` then we need to set the source filter to `user`, but
  // we do this by deriving from props rather than calling setPatternSourceFilter otherwise
  // the user may be confused when switching to another category if the haven't explicitly set
  // this filter themselves.
  const currentPatternSourceFilter = category.name === _utils.myPatternsCategory.name ? _utils.INSERTER_PATTERN_TYPES.user : patternSourceFilter;

  // We need to disable the sync filter option if the source filter is not 'all' or 'user'
  // otherwise applying them will just result in no patterns being shown.
  const shouldDisableSyncFilter = getShouldDisableSyncFilter(currentPatternSourceFilter);

  // We also hide the directory and theme source filter if the category is `myPatterns`
  // otherwise there will only be one option available.
  const shouldHideSourcesFilter = getShouldHideSourcesFilter(category);
  const patternSyncMenuOptions = (0, _element.useMemo)(() => [{
    value: 'all',
    label: (0, _i18n._x)('All', 'patterns')
  }, {
    value: _utils.INSERTER_SYNC_TYPES.full,
    label: (0, _i18n._x)('Synced', 'patterns'),
    disabled: shouldDisableSyncFilter
  }, {
    value: _utils.INSERTER_SYNC_TYPES.unsynced,
    label: (0, _i18n._x)('Not synced', 'patterns'),
    disabled: shouldDisableSyncFilter
  }], [shouldDisableSyncFilter]);
  function handleSetSourceFilterChange(newSourceFilter) {
    setPatternSourceFilter(newSourceFilter);
    if (getShouldDisableSyncFilter(newSourceFilter)) {
      setPatternSyncFilter('all');
    }
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.DropdownMenu, {
      popoverProps: {
        placement: 'right-end'
      },
      label: (0, _i18n.__)('Filter patterns'),
      toggleProps: {
        size: 'compact'
      },
      icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_icons.Icon, {
        icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SVG, {
          width: "24",
          height: "24",
          viewBox: "0 0 24 24",
          fill: "none",
          xmlns: "http://www.w3.org/2000/svg",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Path, {
            d: "M10 17.5H14V16H10V17.5ZM6 6V7.5H18V6H6ZM8 12.5H16V11H8V12.5Z",
            fill: "currentColor"
          })
        })
      }),
      children: () => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [!shouldHideSourcesFilter && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.MenuGroup, {
          label: (0, _i18n.__)('Source'),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.MenuItemsChoice, {
            choices: PATTERN_SOURCE_MENU_OPTIONS,
            onSelect: value => {
              handleSetSourceFilterChange(value);
              scrollContainerRef.current?.scrollTo(0, 0);
            },
            value: currentPatternSourceFilter
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.MenuGroup, {
          label: (0, _i18n.__)('Type'),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.MenuItemsChoice, {
            choices: patternSyncMenuOptions,
            onSelect: value => {
              setPatternSyncFilter(value);
              scrollContainerRef.current?.scrollTo(0, 0);
            },
            value: patternSyncFilter
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "block-editor-inserter__patterns-filter-help",
          children: (0, _element.createInterpolateElement)((0, _i18n.__)('Patterns are available from the <Link>WordPress.org Pattern Directory</Link>, bundled in the active theme, or created by users on this site. Only patterns created on this site can be synced.'), {
            Link: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ExternalLink, {
              href: (0, _i18n.__)('https://wordpress.org/patterns/')
            })
          })
        })]
      })
    })
  });
}
