// @wordpress/block-library/build/block/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = exports.init = void 0;
var _icons = require("@wordpress/icons");
var _coreData = require("@wordpress/core-data");
var _data = require("@wordpress/data");
var _htmlEntities = require("@wordpress/html-entities");
var _initBlock = _interopRequireDefault(require("../utils/init-block"));
var _edit = _interopRequireDefault(require("./edit"));
var _deprecated = _interopRequireDefault(require("./deprecated"));
/**
 * WordPress dependencies
 */
/**
 * Internal dependencies
 */
const metadata = exports.metadata = {
  $schema: "https://schemas.wp.org/trunk/block.json",
  apiVersion: 3,
  name: "core/block",
  title: "Pattern",
  category: "reusable",
  description: "Reuse this design across your site.",
  keywords: ["reusable"],
  textdomain: "default",
  attributes: {
    ref: {
      type: "number"
    },
    content: {
      type: "object",
      "default": {}
    }
  },
  providesContext: {
    "pattern/overrides": "content"
  },
  supports: {
    customClassName: false,
    html: false,
    inserter: false,
    renaming: false,
    interactivity: {
      clientNavigation: true
    }
  }
};
const {
  name
} = metadata;
exports.name = name;
const settings = exports.settings = {
  deprecated: _deprecated.default,
  edit: _edit.default,
  icon: _icons.symbol,
  __experimentalLabel: ({
    ref
  }) => {
    if (!ref) {
      return;
    }
    const entity = (0, _data.select)(_coreData.store).getEditedEntityRecord('postType', 'wp_block', ref);
    if (!entity?.title) {
      return;
    }
    return (0, _htmlEntities.decodeEntities)(entity.title);
  }
};
const init = () => (0, _initBlock.default)({
  name,
  metadata,
  settings
});
exports.init = init;
