// @wordpress/media-utils/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  uploadMedia: true,
  transformAttachment: true,
  validateFileSize: true,
  validateMimeType: true,
  validateMimeTypeForUser: true,
  privateApis: true
};
Object.defineProperty(exports, "privateApis", {
  enumerable: true,
  get: function () {
    return _privateApis.privateApis;
  }
});
Object.defineProperty(exports, "transformAttachment", {
  enumerable: true,
  get: function () {
    return _transformAttachment.transformAttachment;
  }
});
Object.defineProperty(exports, "uploadMedia", {
  enumerable: true,
  get: function () {
    return _uploadMedia.uploadMedia;
  }
});
Object.defineProperty(exports, "validateFileSize", {
  enumerable: true,
  get: function () {
    return _validateFileSize.validateFileSize;
  }
});
Object.defineProperty(exports, "validateMimeType", {
  enumerable: true,
  get: function () {
    return _validateMimeType.validateMimeType;
  }
});
Object.defineProperty(exports, "validateMimeTypeForUser", {
  enumerable: true,
  get: function () {
    return _validateMimeTypeForUser.validateMimeTypeForUser;
  }
});
var _components = require("./components");
Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _components[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _components[key];
    }
  });
});
var _uploadMedia = require("./utils/upload-media");
var _transformAttachment = require("./utils/transform-attachment");
var _validateFileSize = require("./utils/validate-file-size");
var _validateMimeType = require("./utils/validate-mime-type");
var _validateMimeTypeForUser = require("./utils/validate-mime-type-for-user");
var _privateApis = require("./private-apis");
