// @wordpress/fields/build/fields/pattern-title/view.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PATTERN_TYPES = void 0;
exports.default = PatternTitleView;
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _patterns = require("@wordpress/patterns");
var _view = require("../title/view");
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

// @ts-ignore

/**
 * Internal dependencies
 */

const {
  PATTERN_TYPES
} = (0, _lockUnlock.unlock)(_patterns.privateApis);
exports.PATTERN_TYPES = PATTERN_TYPES;
function PatternTitleView({
  item
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_view.BaseTitleView, {
    item: item,
    className: "fields-field__pattern-title",
    children: item.type === PATTERN_TYPES.theme && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Tooltip, {
      placement: "top",
      text: (0, _i18n.__)('This pattern cannot be edited.'),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_icons.Icon, {
        icon: _icons.lockSmall,
        size: 24
      })
    })
  });
}
