// @wordpress/block-editor/build/components/colors-gradients/panel-color-gradient-settings.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PanelColorGradientSettings;
var _native = require("@react-navigation/native");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _blockSettings = require("../block-settings");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PanelColorGradientSettings({
  settings,
  title,
  children
}) {
  const navigation = (0, _native.useNavigation)();
  const mappedSettings = (0, _element.useMemo)(() => {
    return settings.map(({
      onColorChange,
      onColorCleared,
      colorValue,
      onGradientChange,
      gradientValue,
      label
    }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.ColorControl, {
      onPress: () => {
        navigation.navigate(_blockSettings.blockSettingsScreens.color, {
          onColorChange,
          colorValue: gradientValue || colorValue,
          gradientValue,
          onGradientChange,
          onColorCleared,
          label
        });
      },
      label: label,
      color: gradientValue || colorValue
    }, `color-setting-${label}`));
  }, [settings]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.PanelBody, {
      title: title,
      children: mappedSettings
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.PanelBody, {
      children: children
    })]
  });
}
