// @wordpress/block-editor/build-module/components/multi-selection-inspector/index.js
/**
 * WordPress dependencies
 */
import { sprintf, _n } from '@wordpress/i18n';
import { useSelect } from '@wordpress/data';
import { copy } from '@wordpress/icons';
import { __experimentalHStack as HStack } from '@wordpress/components';

/**
 * Internal dependencies
 */
import BlockIcon from '../block-icon';
import { store as blockEditorStore } from '../../store';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export default function MultiSelectionInspector() {
  const selectedBlockCount = useSelect(select => select(blockEditorStore).getSelectedBlockCount(), []);
  return /*#__PURE__*/_jsxs(HStack, {
    justify: "flex-start",
    spacing: 2,
    className: "block-editor-multi-selection-inspector__card",
    children: [/*#__PURE__*/_jsx(BlockIcon, {
      icon: copy,
      showColors: true
    }), /*#__PURE__*/_jsx("div", {
      className: "block-editor-multi-selection-inspector__card-title",
      children: sprintf(/* translators: %d: number of blocks */
      _n('%d Block', '%d Blocks', selectedBlockCount), selectedBlockCount)
    })]
  });
}
