// @wordpress/fields/build/fields/featured-image/featured-image-edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FeaturedImageEdit = void 0;
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _mediaUtils = require("@wordpress/media-utils");
var _icons = require("@wordpress/icons");
var _coreData = require("@wordpress/core-data");
var _i18n = require("@wordpress/i18n");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

// @ts-ignore

/**
 * Internal dependencies
 */

const FeaturedImageEdit = ({
  data,
  field,
  onChange
}) => {
  const {
    id
  } = field;
  const value = field.getValue({
    item: data
  });
  const media = (0, _data.useSelect)(select => {
    const {
      getEntityRecord
    } = select(_coreData.store);
    return getEntityRecord('root', 'media', value);
  }, [value]);
  const onChangeControl = (0, _element.useCallback)(newValue => onChange({
    [id]: newValue
  }), [id, onChange]);
  const url = media?.source_url;
  const title = media?.title?.rendered;
  const ref = (0, _element.useRef)(null);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("fieldset", {
    className: "fields-controls__featured-image",
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "fields-controls__featured-image-container",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_mediaUtils.MediaUpload, {
        onSelect: selectedMedia => {
          onChangeControl(selectedMedia.id);
        },
        allowedTypes: ['image'],
        render: ({
          open
        }) => {
          return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            ref: ref,
            role: "button",
            tabIndex: -1,
            onClick: () => {
              open();
            },
            onKeyDown: open,
            children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalGrid, {
              rowGap: 0,
              columnGap: 8,
              templateColumns: "24px 1fr 24px",
              children: [url && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
                children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
                  className: "fields-controls__featured-image-image",
                  alt: "",
                  width: 24,
                  height: 24,
                  src: url
                }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                  className: "fields-controls__featured-image-title",
                  children: title
                })]
              }), !url && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
                children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                  className: "fields-controls__featured-image-placeholder",
                  style: {
                    width: '24px',
                    height: '24px'
                  }
                }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                  className: "fields-controls__featured-image-title",
                  children: (0, _i18n.__)('Choose an image…')
                })]
              }), url && /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
                  size: "small",
                  className: "fields-controls__featured-image-remove-button",
                  icon: _icons.lineSolid,
                  onClick: event => {
                    event.stopPropagation();
                    onChangeControl(0);
                  }
                })
              })]
            })
          });
        }
      })
    })
  });
};
exports.FeaturedImageEdit = FeaturedImageEdit;
