// @wordpress/block-library/build/query-total/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = exports.init = void 0;
var _edit = _interopRequireDefault(require("./edit"));
var _initBlock = _interopRequireDefault(require("../utils/init-block"));
var _icons = require("./icons");
/**
 * Internal dependencies
 */
const metadata = exports.metadata = {
  $schema: "https://schemas.wp.org/trunk/block.json",
  apiVersion: 3,
  name: "core/query-total",
  title: "Query Total",
  category: "theme",
  ancestor: ["core/query"],
  description: "Display the total number of results in a query.",
  textdomain: "default",
  attributes: {
    displayType: {
      type: "string",
      "default": "total-results"
    }
  },
  usesContext: ["queryId", "query"],
  supports: {
    align: ["wide", "full"],
    html: false,
    spacing: {
      margin: true,
      padding: true
    },
    color: {
      gradients: true,
      __experimentalDefaultControls: {
        background: true,
        text: true
      }
    },
    typography: {
      fontSize: true,
      lineHeight: true,
      __experimentalFontFamily: true,
      __experimentalFontWeight: true,
      __experimentalFontStyle: true,
      __experimentalTextTransform: true,
      __experimentalTextDecoration: true,
      __experimentalLetterSpacing: true,
      __experimentalDefaultControls: {
        fontSize: true
      }
    },
    __experimentalBorder: {
      radius: true,
      color: true,
      width: true,
      style: true,
      __experimentalDefaultControls: {
        radius: true,
        color: true,
        width: true,
        style: true
      }
    },
    interactivity: {
      clientNavigation: true
    }
  },
  style: "wp-block-query-total"
};
/* Block settings */
const {
  name
} = metadata;
exports.name = name;
const settings = exports.settings = {
  icon: _icons.queryTotal,
  edit: _edit.default
};
const init = () => (0, _initBlock.default)({
  name,
  metadata,
  settings
});
exports.init = init;
