// @wordpress/block-library/build/comments-pagination-numbers/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CommentsPaginationNumbersEdit;
var _blockEditor = require("@wordpress/block-editor");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

const PaginationItem = ({
  content,
  tag: Tag = 'a',
  extraClass = ''
}) => Tag === 'a' ? /*#__PURE__*/(0, _jsxRuntime.jsx)(Tag, {
  className: `page-numbers ${extraClass}`,
  href: "#comments-pagination-numbers-pseudo-link",
  onClick: event => event.preventDefault(),
  children: content
}) : /*#__PURE__*/(0, _jsxRuntime.jsx)(Tag, {
  className: `page-numbers ${extraClass}`,
  children: content
});
function CommentsPaginationNumbersEdit() {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    ...(0, _blockEditor.useBlockProps)(),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "1"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "2"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "3",
      tag: "span",
      extraClass: "current"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "4"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "5"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "...",
      tag: "span",
      extraClass: "dots"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PaginationItem, {
      content: "8"
    })]
  });
}
