// @wordpress/edit-post/build/components/keyboard-shortcuts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _keyboardShortcuts = require("@wordpress/keyboard-shortcuts");
var _i18n = require("@wordpress/i18n");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function KeyboardShortcuts() {
  const {
    toggleFullscreenMode
  } = (0, _data.useDispatch)(_store.store);
  const {
    registerShortcut
  } = (0, _data.useDispatch)(_keyboardShortcuts.store);
  (0, _element.useEffect)(() => {
    registerShortcut({
      name: 'core/edit-post/toggle-fullscreen',
      category: 'global',
      description: (0, _i18n.__)('Enable or disable fullscreen mode.'),
      keyCombination: {
        modifier: 'secondary',
        character: 'f'
      }
    });
  }, []);
  (0, _keyboardShortcuts.useShortcut)('core/edit-post/toggle-fullscreen', () => {
    toggleFullscreenMode();
  });
  return null;
}
var _default = exports.default = KeyboardShortcuts;
