// @wordpress/block-editor/build/components/block-switcher/preview-block-popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreviewBlockPopover;
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _compose = require("@wordpress/compose");
var _blockPreview = _interopRequireDefault(require("../block-preview"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PreviewBlockPopover({
  blocks
}) {
  const isMobile = (0, _compose.useViewportMatch)('medium', '<');
  if (isMobile) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "block-editor-block-switcher__popover-preview-container",
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Popover, {
      className: "block-editor-block-switcher__popover-preview",
      placement: "right-start",
      focusOnMount: false,
      offset: 16,
      children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "block-editor-block-switcher__preview",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "block-editor-block-switcher__preview-title",
          children: (0, _i18n.__)('Preview')
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockPreview.default, {
          viewportWidth: 601,
          blocks: blocks
        })]
      })
    })
  });
}
