// @wordpress/block-library/build/navigation/edit/deleted-navigation-warning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _blockEditor = require("@wordpress/block-editor");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

function DeletedNavigationWarning({
  onCreateNew,
  isNotice = false
}) {
  const [isButtonDisabled, setIsButtonDisabled] = (0, _element.useState)(false);
  const handleButtonClick = () => {
    setIsButtonDisabled(true);
    onCreateNew();
  };
  const message = (0, _element.createInterpolateElement)((0, _i18n.__)('Navigation Menu has been deleted or is unavailable. <button>Create a new Menu?</button>'), {
    button: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
      __next40pxDefaultSize: true,
      onClick: handleButtonClick,
      variant: "link",
      disabled: isButtonDisabled,
      accessibleWhenDisabled: true
    })
  });
  return isNotice ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Notice, {
    status: "warning",
    isDismissible: false,
    children: message
  }) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.Warning, {
    children: message
  });
}
var _default = exports.default = DeletedNavigationWarning;
