// @wordpress/upload-media/build/store/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Type = exports.OperationType = exports.ItemStatus = void 0;
let Type = exports.Type = /*#__PURE__*/function (Type) {
  Type["Unknown"] = "REDUX_UNKNOWN";
  Type["Add"] = "ADD_ITEM";
  Type["Prepare"] = "PREPARE_ITEM";
  Type["Cancel"] = "CANCEL_ITEM";
  Type["Remove"] = "REMOVE_ITEM";
  Type["PauseItem"] = "PAUSE_ITEM";
  Type["ResumeItem"] = "RESUME_ITEM";
  Type["PauseQueue"] = "PAUSE_QUEUE";
  Type["ResumeQueue"] = "RESUME_QUEUE";
  Type["OperationStart"] = "OPERATION_START";
  Type["OperationFinish"] = "OPERATION_FINISH";
  Type["AddOperations"] = "ADD_OPERATIONS";
  Type["CacheBlobUrl"] = "CACHE_BLOB_URL";
  Type["RevokeBlobUrls"] = "REVOKE_BLOB_URLS";
  Type["UpdateSettings"] = "UPDATE_SETTINGS";
  return Type;
}({}); // Must match the Attachment type from the media-utils package.
let ItemStatus = exports.ItemStatus = /*#__PURE__*/function (ItemStatus) {
  ItemStatus["Processing"] = "PROCESSING";
  ItemStatus["Paused"] = "PAUSED";
  return ItemStatus;
}({});
let OperationType = exports.OperationType = /*#__PURE__*/function (OperationType) {
  OperationType["Prepare"] = "PREPARE";
  OperationType["Upload"] = "UPLOAD";
  return OperationType;
}({});
