// @wordpress/block-editor/build/components/multi-selection-inspector/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MultiSelectionInspector;
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _blockIcon = _interopRequireDefault(require("../block-icon"));
var _store = require("../../store");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function MultiSelectionInspector() {
  const selectedBlockCount = (0, _data.useSelect)(select => select(_store.store).getSelectedBlockCount(), []);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    justify: "flex-start",
    spacing: 2,
    className: "block-editor-multi-selection-inspector__card",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_blockIcon.default, {
      icon: _icons.copy,
      showColors: true
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "block-editor-multi-selection-inspector__card-title",
      children: (0, _i18n.sprintf)(/* translators: %d: number of blocks */
      (0, _i18n._n)('%d Block', '%d Blocks', selectedBlockCount), selectedBlockCount)
    })]
  });
}
