// @wordpress/editor/build/components/editor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _constants = require("../../store/constants");
var _editorInterface = _interopRequireDefault(require("../editor-interface"));
var _provider = require("../provider");
var _sidebar = _interopRequireDefault(require("../sidebar"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function Editor({
  postType,
  postId,
  templateId,
  settings,
  children,
  initialEdits,
  // This could be part of the settings.
  onActionPerformed,
  // The following abstractions are not ideal but necessary
  // to account for site editor and post editor differences for now.
  extraContent,
  extraSidebarPanels,
  ...props
}) {
  const {
    post,
    template,
    hasLoadedPost,
    error
  } = (0, _data.useSelect)(select => {
    const {
      getEntityRecord,
      getResolutionError,
      hasFinishedResolution
    } = select(_coreData.store);
    const postArgs = ['postType', postType, postId];
    return {
      post: getEntityRecord(...postArgs),
      template: templateId ? getEntityRecord('postType', _constants.TEMPLATE_POST_TYPE, templateId) : undefined,
      hasLoadedPost: hasFinishedResolution('getEntityRecord', postArgs),
      error: getResolutionError('getEntityRecord', postArgs)?.message
    };
  }, [postType, postId, templateId]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [hasLoadedPost && !post && /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Notice, {
      status: !!error ? 'error' : 'warning',
      isDismissible: false,
      children: !error ? (0, _i18n.__)("You attempted to edit an item that doesn't exist. Perhaps it was deleted?") : error
    }), !!post && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_provider.ExperimentalEditorProvider, {
      post: post,
      __unstableTemplate: template,
      settings: settings,
      initialEdits: initialEdits,
      useSubRegistry: false,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_editorInterface.default, {
        ...props,
        children: extraContent
      }), children, /*#__PURE__*/(0, _jsxRuntime.jsx)(_sidebar.default, {
        onActionPerformed: onActionPerformed,
        extraPanels: extraSidebarPanels
      })]
    })]
  });
}
var _default = exports.default = Editor;
