// @wordpress/dataviews/build/dataform-controls/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getControl = getControl;
exports.getControlByType = getControlByType;
var _datetime = _interopRequireDefault(require("./datetime"));
var _integer = _interopRequireDefault(require("./integer"));
var _radio = _interopRequireDefault(require("./radio"));
var _select = _interopRequireDefault(require("./select"));
var _text = _interopRequireDefault(require("./text"));
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

const FORM_CONTROLS = {
  datetime: _datetime.default,
  integer: _integer.default,
  radio: _radio.default,
  select: _select.default,
  text: _text.default
};
function getControl(field, fieldTypeDefinition) {
  if (typeof field.Edit === 'function') {
    return field.Edit;
  }
  if (typeof field.Edit === 'string') {
    return getControlByType(field.Edit);
  }
  if (field.elements) {
    return getControlByType('select');
  }
  if (typeof fieldTypeDefinition.Edit === 'string') {
    return getControlByType(fieldTypeDefinition.Edit);
  }
  return fieldTypeDefinition.Edit;
}
function getControlByType(type) {
  if (Object.keys(FORM_CONTROLS).includes(type)) {
    return FORM_CONTROLS[type];
  }
  throw 'Control ' + type + ' not found';
}
