// @wordpress/editor/build/store/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TEMPLATE_POST_TYPES = exports.TEMPLATE_POST_TYPE = exports.TEMPLATE_PART_POST_TYPE = exports.TEMPLATE_PART_AREA_DEFAULT_CATEGORY = exports.TEMPLATE_ORIGINS = exports.STORE_NAME = exports.PERMALINK_POSTNAME_REGEX = exports.PATTERN_POST_TYPE = exports.ONE_MINUTE_IN_MS = exports.NAVIGATION_POST_TYPE = exports.GLOBAL_POST_TYPES = exports.EDIT_MERGE_PROPERTIES = exports.AUTOSAVE_PROPERTIES = void 0;
/**
 * Set of post properties for which edits should assume a merging behavior,
 * assuming an object value.
 *
 * @type {Set}
 */
const EDIT_MERGE_PROPERTIES = exports.EDIT_MERGE_PROPERTIES = new Set(['meta']);

/**
 * Constant for the store module (or reducer) key.
 */
const STORE_NAME = exports.STORE_NAME = 'core/editor';
const PERMALINK_POSTNAME_REGEX = exports.PERMALINK_POSTNAME_REGEX = /%(?:postname|pagename)%/;
const ONE_MINUTE_IN_MS = exports.ONE_MINUTE_IN_MS = 60 * 1000;
const AUTOSAVE_PROPERTIES = exports.AUTOSAVE_PROPERTIES = ['title', 'excerpt', 'content'];
const TEMPLATE_PART_AREA_DEFAULT_CATEGORY = exports.TEMPLATE_PART_AREA_DEFAULT_CATEGORY = 'uncategorized';
const TEMPLATE_POST_TYPE = exports.TEMPLATE_POST_TYPE = 'wp_template';
const TEMPLATE_PART_POST_TYPE = exports.TEMPLATE_PART_POST_TYPE = 'wp_template_part';
const PATTERN_POST_TYPE = exports.PATTERN_POST_TYPE = 'wp_block';
const NAVIGATION_POST_TYPE = exports.NAVIGATION_POST_TYPE = 'wp_navigation';
const TEMPLATE_ORIGINS = exports.TEMPLATE_ORIGINS = {
  custom: 'custom',
  theme: 'theme',
  plugin: 'plugin'
};
const TEMPLATE_POST_TYPES = exports.TEMPLATE_POST_TYPES = ['wp_template', 'wp_template_part'];
const GLOBAL_POST_TYPES = exports.GLOBAL_POST_TYPES = [...TEMPLATE_POST_TYPES, 'wp_block', 'wp_navigation'];
