// @wordpress/components/build/menu/submenu-trigger-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SubmenuTriggerItem = void 0;
var Ariakit = _interopRequireWildcard(require("@ariakit/react"));
var _element = require("@wordpress/element");
var _icons = require("@wordpress/icons");
var _context = require("./context");
var _item = require("./item");
var Styled = _interopRequireWildcard(require("./styles"));
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const SubmenuTriggerItem = exports.SubmenuTriggerItem = (0, _element.forwardRef)(function SubmenuTriggerItem({
  suffix,
  ...otherProps
}, ref) {
  const menuContext = (0, _element.useContext)(_context.Context);
  if (!menuContext?.store.parent) {
    throw new Error('Menu.SubmenuTriggerItem can only be rendered inside a nested Menu component');
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Ariakit.MenuButton, {
    ref: ref,
    accessibleWhenDisabled: true,
    store: menuContext.store,
    render: /*#__PURE__*/(0, _jsxRuntime.jsx)(_item.Item, {
      ...otherProps,
      // The menu item needs to register and be part of the parent menu.
      // Without specifying the store explicitly, the `Item` component
      // would otherwise read the store via context and pick up the one from
      // the sub-menu `Menu` component.
      store: menuContext.store.parent,
      suffix: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [suffix, /*#__PURE__*/(0, _jsxRuntime.jsx)(Styled.SubmenuChevronIcon, {
          "aria-hidden": "true",
          icon: _icons.chevronRightSmall,
          size: 24,
          preserveAspectRatio: "xMidYMid slice"
        })]
      })
    })
  });
});
