// @wordpress/core-data/build/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ALLOWED_RESOURCE_ACTIONS", {
  enumerable: true,
  get: function () {
    return _userPermissions.ALLOWED_RESOURCE_ACTIONS;
  }
});
Object.defineProperty(exports, "RECEIVE_INTERMEDIATE_RESULTS", {
  enumerable: true,
  get: function () {
    return _receiveIntermediateResults.RECEIVE_INTERMEDIATE_RESULTS;
  }
});
Object.defineProperty(exports, "conservativeMapItem", {
  enumerable: true,
  get: function () {
    return _conservativeMapItem.default;
  }
});
Object.defineProperty(exports, "forwardResolver", {
  enumerable: true,
  get: function () {
    return _forwardResolver.default;
  }
});
Object.defineProperty(exports, "getNestedValue", {
  enumerable: true,
  get: function () {
    return _getNestedValue.default;
  }
});
Object.defineProperty(exports, "getNormalizedCommaSeparable", {
  enumerable: true,
  get: function () {
    return _getNormalizedCommaSeparable.default;
  }
});
Object.defineProperty(exports, "getUserPermissionCacheKey", {
  enumerable: true,
  get: function () {
    return _userPermissions.getUserPermissionCacheKey;
  }
});
Object.defineProperty(exports, "getUserPermissionsFromAllowHeader", {
  enumerable: true,
  get: function () {
    return _userPermissions.getUserPermissionsFromAllowHeader;
  }
});
Object.defineProperty(exports, "ifMatchingAction", {
  enumerable: true,
  get: function () {
    return _ifMatchingAction.default;
  }
});
Object.defineProperty(exports, "isNumericID", {
  enumerable: true,
  get: function () {
    return _isNumericId.default;
  }
});
Object.defineProperty(exports, "isRawAttribute", {
  enumerable: true,
  get: function () {
    return _isRawAttribute.default;
  }
});
Object.defineProperty(exports, "onSubKey", {
  enumerable: true,
  get: function () {
    return _onSubKey.default;
  }
});
Object.defineProperty(exports, "replaceAction", {
  enumerable: true,
  get: function () {
    return _replaceAction.default;
  }
});
Object.defineProperty(exports, "setNestedValue", {
  enumerable: true,
  get: function () {
    return _setNestedValue.default;
  }
});
Object.defineProperty(exports, "withWeakMapCache", {
  enumerable: true,
  get: function () {
    return _withWeakMapCache.default;
  }
});
var _conservativeMapItem = _interopRequireDefault(require("./conservative-map-item"));
var _getNormalizedCommaSeparable = _interopRequireDefault(require("./get-normalized-comma-separable"));
var _ifMatchingAction = _interopRequireDefault(require("./if-matching-action"));
var _forwardResolver = _interopRequireDefault(require("./forward-resolver"));
var _onSubKey = _interopRequireDefault(require("./on-sub-key"));
var _replaceAction = _interopRequireDefault(require("./replace-action"));
var _withWeakMapCache = _interopRequireDefault(require("./with-weak-map-cache"));
var _isRawAttribute = _interopRequireDefault(require("./is-raw-attribute"));
var _setNestedValue = _interopRequireDefault(require("./set-nested-value"));
var _getNestedValue = _interopRequireDefault(require("./get-nested-value"));
var _isNumericId = _interopRequireDefault(require("./is-numeric-id"));
var _userPermissions = require("./user-permissions");
var _receiveIntermediateResults = require("./receive-intermediate-results");
