// @wordpress/dataviews/build/dataviews-layouts/table/column-header-menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _lockUnlock = require("../../lock-unlock");
var _utils = require("../../utils");
var _constants = require("../../constants");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Menu
} = (0, _lockUnlock.unlock)(_components.privateApis);
function WithMenuSeparators({
  children
}) {
  return _element.Children.toArray(children).filter(Boolean).map((child, i) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_element.Fragment, {
    children: [i > 0 && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Separator, {}), child]
  }, i));
}
const _HeaderMenu = (0, _element.forwardRef)(function HeaderMenu({
  fieldId,
  view,
  fields,
  onChangeView,
  onHide,
  setOpenedFilter,
  canMove = true
}, ref) {
  var _view$fields;
  const visibleFieldIds = (_view$fields = view.fields) !== null && _view$fields !== void 0 ? _view$fields : [];
  const index = visibleFieldIds?.indexOf(fieldId);
  const isSorted = view.sort?.field === fieldId;
  let isHidable = false;
  let isSortable = false;
  let canAddFilter = false;
  let operators = [];
  const field = fields.find(f => f.id === fieldId);
  if (!field) {
    // No combined or regular field found.
    return null;
  }
  isHidable = field.enableHiding !== false;
  isSortable = field.enableSorting !== false;
  const header = field.header;
  operators = (0, _utils.sanitizeOperators)(field);
  // Filter can be added:
  // 1. If the field is not already part of a view's filters.
  // 2. If the field meets the type and operator requirements.
  // 3. If it's not primary. If it is, it should be already visible.
  canAddFilter = !view.filters?.some(_filter => fieldId === _filter.field) && !!field.elements?.length && !!operators.length && !field.filterBy?.isPrimary;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Menu, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(Menu.TriggerButton, {
      render: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
        size: "compact",
        className: "dataviews-view-table-header-button",
        ref: ref,
        variant: "tertiary"
      }),
      children: [header, view.sort && isSorted && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        "aria-hidden": "true",
        children: _constants.sortArrows[view.sort.direction]
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Popover, {
      style: {
        minWidth: '240px'
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(WithMenuSeparators, {
        children: [isSortable && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Group, {
          children: _constants.SORTING_DIRECTIONS.map(direction => {
            const isChecked = view.sort && isSorted && view.sort.direction === direction;
            const value = `${fieldId}-${direction}`;
            return /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.RadioItem, {
              // All sorting radio items share the same name, so that
              // selecting a sorting option automatically deselects the
              // previously selected one, even if it is displayed in
              // another submenu. The field and direction are passed via
              // the `value` prop.
              name: "view-table-sorting",
              value: value,
              checked: isChecked,
              onChange: () => {
                onChangeView({
                  ...view,
                  sort: {
                    field: fieldId,
                    direction
                  },
                  showLevels: false
                });
              },
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.ItemLabel, {
                children: _constants.sortLabels[direction]
              })
            }, value);
          })
        }), canAddFilter && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Group, {
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Item, {
            prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
              icon: _icons.funnel
            }),
            onClick: () => {
              setOpenedFilter(fieldId);
              onChangeView({
                ...view,
                page: 1,
                filters: [...(view.filters || []), {
                  field: fieldId,
                  value: undefined,
                  operator: operators[0]
                }]
              });
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.ItemLabel, {
              children: (0, _i18n.__)('Add filter')
            })
          })
        }), (canMove || isHidable) && field && /*#__PURE__*/(0, _jsxRuntime.jsxs)(Menu.Group, {
          children: [canMove && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Item, {
            prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
              icon: _icons.arrowLeft
            }),
            disabled: index < 1,
            onClick: () => {
              var _visibleFieldIds$slic;
              onChangeView({
                ...view,
                fields: [...((_visibleFieldIds$slic = visibleFieldIds.slice(0, index - 1)) !== null && _visibleFieldIds$slic !== void 0 ? _visibleFieldIds$slic : []), fieldId, visibleFieldIds[index - 1], ...visibleFieldIds.slice(index + 1)]
              });
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.ItemLabel, {
              children: (0, _i18n.__)('Move left')
            })
          }), canMove && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Item, {
            prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
              icon: _icons.arrowRight
            }),
            disabled: index >= visibleFieldIds.length - 1,
            onClick: () => {
              var _visibleFieldIds$slic2;
              onChangeView({
                ...view,
                fields: [...((_visibleFieldIds$slic2 = visibleFieldIds.slice(0, index)) !== null && _visibleFieldIds$slic2 !== void 0 ? _visibleFieldIds$slic2 : []), visibleFieldIds[index + 1], fieldId, ...visibleFieldIds.slice(index + 2)]
              });
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.ItemLabel, {
              children: (0, _i18n.__)('Move right')
            })
          }), isHidable && field && /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.Item, {
            prefix: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
              icon: _icons.unseen
            }),
            onClick: () => {
              onHide(field);
              onChangeView({
                ...view,
                fields: visibleFieldIds.filter(id => id !== fieldId)
              });
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Menu.ItemLabel, {
              children: (0, _i18n.__)('Hide column')
            })
          })]
        })]
      })
    })]
  });
});

// @ts-expect-error Lift the `Item` type argument through the forwardRef.
const ColumnHeaderMenu = _HeaderMenu;
var _default = exports.default = ColumnHeaderMenu;
