// @wordpress/dataviews/build/dataviews-layouts/table/column-primary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _getClickableItemProps = _interopRequireDefault(require("../utils/get-clickable-item-props"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ColumnPrimary({
  item,
  level,
  titleField,
  mediaField,
  descriptionField,
  onClickItem,
  isItemClickable
}) {
  const clickableProps = (0, _getClickableItemProps.default)({
    item,
    isItemClickable,
    onClickItem,
    className: 'dataviews-view-table__cell-content-wrapper dataviews-title-field'
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    spacing: 3,
    justify: "flex-start",
    children: [mediaField && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "dataviews-view-table__cell-content-wrapper dataviews-column-primary__media",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(mediaField.render, {
        item: item
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalVStack, {
      spacing: 0,
      children: [titleField && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        ...clickableProps,
        children: [level !== undefined && /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
          className: "dataviews-view-table__level",
          children: ['—'.repeat(level), "\xA0"]
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(titleField.render, {
          item: item
        })]
      }), descriptionField && /*#__PURE__*/(0, _jsxRuntime.jsx)(descriptionField.render, {
        item: item
      })]
    })]
  });
}
var _default = exports.default = ColumnPrimary;
