// @wordpress/fields/build-module/components/create-template-part-modal/index.js
/**
 * WordPress dependencies
 */
import { Icon, BaseControl, TextControl, Button, Modal, __experimentalHStack as HStack, __experimentalVStack as VStack } from '@wordpress/components';
import { useInstanceId } from '@wordpress/compose';
import { store as coreStore } from '@wordpress/core-data';
import { useDispatch, useSelect } from '@wordpress/data';
import { useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { check, footer as footerIcon, header as headerIcon, sidebar as sidebarIcon, symbolFilled as symbolFilledIcon } from '@wordpress/icons';
import { store as noticesStore } from '@wordpress/notices';
// @ts-expect-error serialize is not typed
import { serialize } from '@wordpress/blocks';

/**
 * Internal dependencies
 */
import { getCleanTemplatePartSlug, getUniqueTemplatePartTitle, useExistingTemplateParts } from './utils';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
function getAreaRadioId(value, instanceId) {
  return `fields-create-template-part-modal__area-option-${value}-${instanceId}`;
}
function getAreaRadioDescriptionId(value, instanceId) {
  return `fields-create-template-part-modal__area-option-description-${value}-${instanceId}`;
}
/**
 * A React component that renders a modal for creating a template part. The modal displays a title and the contents for creating the template part.
 * This component should not live in this package, it should be moved to a dedicated package responsible for managing template.
 * @param {Object} props            The component props.
 * @param          props.modalTitle
 */
export default function CreateTemplatePartModal({
  modalTitle,
  ...restProps
}) {
  const defaultModalTitle = useSelect(select => select(coreStore).getPostType('wp_template_part')?.labels?.add_new_item, []);
  return /*#__PURE__*/_jsx(Modal, {
    title: modalTitle || defaultModalTitle,
    onRequestClose: restProps.closeModal,
    overlayClassName: "fields-create-template-part-modal",
    focusOnMount: "firstContentElement",
    size: "medium",
    children: /*#__PURE__*/_jsx(CreateTemplatePartModalContents, {
      ...restProps
    })
  });
}
const getTemplatePartIcon = iconName => {
  if ('header' === iconName) {
    return headerIcon;
  } else if ('footer' === iconName) {
    return footerIcon;
  } else if ('sidebar' === iconName) {
    return sidebarIcon;
  }
  return symbolFilledIcon;
};

/**
 * A React component that renders the content of a model for creating a template part.
 * This component should not live in this package; it should be moved to a dedicated package responsible for managing template.
 *
 * @param {Object}   props                             - The component props.
 * @param {string}   [props.defaultArea=uncategorized] - The default area for the template part.
 * @param {Array}    [props.blocks=[]]                 - The blocks to be included in the template part.
 * @param {string}   [props.confirmLabel='Add']        - The label for the confirm button.
 * @param {Function} props.closeModal                  - Function to close the modal.
 * @param {Function} props.onCreate                    - Function to call when the template part is successfully created.
 * @param {Function} [props.onError]                   - Function to call when there is an error creating the template part.
 * @param {string}   [props.defaultTitle='']           - The default title for the template part.
 */
export function CreateTemplatePartModalContents({
  defaultArea = 'uncategorized',
  blocks = [],
  confirmLabel = __('Add'),
  closeModal,
  onCreate,
  onError,
  defaultTitle = ''
}) {
  const {
    createErrorNotice
  } = useDispatch(noticesStore);
  const {
    saveEntityRecord
  } = useDispatch(coreStore);
  const existingTemplateParts = useExistingTemplateParts();
  const [title, setTitle] = useState(defaultTitle);
  const [area, setArea] = useState(defaultArea);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const instanceId = useInstanceId(CreateTemplatePartModal);
  const defaultTemplatePartAreas = useSelect(select => select(coreStore).getCurrentTheme()?.default_template_part_areas, []);
  async function createTemplatePart() {
    if (!title || isSubmitting) {
      return;
    }
    try {
      setIsSubmitting(true);
      const uniqueTitle = getUniqueTemplatePartTitle(title, existingTemplateParts);
      const cleanSlug = getCleanTemplatePartSlug(uniqueTitle);
      const templatePart = await saveEntityRecord('postType', 'wp_template_part', {
        slug: cleanSlug,
        title: uniqueTitle,
        content: serialize(blocks),
        area
      }, {
        throwOnError: true
      });
      await onCreate(templatePart);

      // TODO: Add a success notice?
    } catch (error) {
      const errorMessage = error instanceof Error && 'code' in error && error.message && error.code !== 'unknown_error' ? error.message : __('An error occurred while creating the template part.');
      createErrorNotice(errorMessage, {
        type: 'snackbar'
      });
      onError?.();
    } finally {
      setIsSubmitting(false);
    }
  }
  return /*#__PURE__*/_jsx("form", {
    onSubmit: async event => {
      event.preventDefault();
      await createTemplatePart();
    },
    children: /*#__PURE__*/_jsxs(VStack, {
      spacing: "4",
      children: [/*#__PURE__*/_jsx(TextControl, {
        __next40pxDefaultSize: true,
        __nextHasNoMarginBottom: true,
        label: __('Name'),
        value: title,
        onChange: setTitle,
        required: true
      }), /*#__PURE__*/_jsxs("fieldset", {
        children: [/*#__PURE__*/_jsx(BaseControl.VisualLabel, {
          as: "legend",
          children: __('Area')
        }), /*#__PURE__*/_jsx("div", {
          className: "fields-create-template-part-modal__area-radio-group",
          children: (defaultTemplatePartAreas !== null && defaultTemplatePartAreas !== void 0 ? defaultTemplatePartAreas : []).map(item => {
            const icon = getTemplatePartIcon(item.icon);
            return /*#__PURE__*/_jsxs("div", {
              className: "fields-create-template-part-modal__area-radio-wrapper",
              children: [/*#__PURE__*/_jsx("input", {
                type: "radio",
                id: getAreaRadioId(item.area, instanceId),
                name: `fields-create-template-part-modal__area-${instanceId}`,
                value: item.area,
                checked: area === item.area,
                onChange: () => {
                  setArea(item.area);
                },
                "aria-describedby": getAreaRadioDescriptionId(item.area, instanceId)
              }), /*#__PURE__*/_jsx(Icon, {
                icon: icon,
                className: "fields-create-template-part-modal__area-radio-icon"
              }), /*#__PURE__*/_jsx("label", {
                htmlFor: getAreaRadioId(item.area, instanceId),
                className: "fields-create-template-part-modal__area-radio-label",
                children: item.label
              }), /*#__PURE__*/_jsx(Icon, {
                icon: check,
                className: "fields-create-template-part-modal__area-radio-checkmark"
              }), /*#__PURE__*/_jsx("p", {
                className: "fields-create-template-part-modal__area-radio-description",
                id: getAreaRadioDescriptionId(item.area, instanceId),
                children: item.description
              })]
            }, item.area);
          })
        })]
      }), /*#__PURE__*/_jsxs(HStack, {
        justify: "right",
        children: [/*#__PURE__*/_jsx(Button, {
          __next40pxDefaultSize: true,
          variant: "tertiary",
          onClick: () => {
            closeModal();
          },
          children: __('Cancel')
        }), /*#__PURE__*/_jsx(Button, {
          __next40pxDefaultSize: true,
          variant: "primary",
          type: "submit",
          "aria-disabled": !title || isSubmitting,
          isBusy: isSubmitting,
          children: confirmLabel
        })]
      })]
    })
  });
}
