// @wordpress/dataviews/build/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sortValues = exports.sortLabels = exports.sortIcons = exports.sortArrows = exports.SORTING_DIRECTIONS = exports.OPERATOR_IS_NOT_ALL = exports.OPERATOR_IS_NOT = exports.OPERATOR_IS_NONE = exports.OPERATOR_IS_ANY = exports.OPERATOR_IS_ALL = exports.OPERATOR_IS = exports.OPERATORS = exports.LAYOUT_TABLE = exports.LAYOUT_LIST = exports.LAYOUT_GRID = exports.ALL_OPERATORS = void 0;
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

// Filter operators.
const OPERATOR_IS = exports.OPERATOR_IS = 'is';
const OPERATOR_IS_NOT = exports.OPERATOR_IS_NOT = 'isNot';
const OPERATOR_IS_ANY = exports.OPERATOR_IS_ANY = 'isAny';
const OPERATOR_IS_NONE = exports.OPERATOR_IS_NONE = 'isNone';
const OPERATOR_IS_ALL = exports.OPERATOR_IS_ALL = 'isAll';
const OPERATOR_IS_NOT_ALL = exports.OPERATOR_IS_NOT_ALL = 'isNotAll';
const ALL_OPERATORS = exports.ALL_OPERATORS = [OPERATOR_IS, OPERATOR_IS_NOT, OPERATOR_IS_ANY, OPERATOR_IS_NONE, OPERATOR_IS_ALL, OPERATOR_IS_NOT_ALL];
const OPERATORS = exports.OPERATORS = {
  [OPERATOR_IS]: {
    key: 'is-filter',
    label: (0, _i18n.__)('Is')
  },
  [OPERATOR_IS_NOT]: {
    key: 'is-not-filter',
    label: (0, _i18n.__)('Is not')
  },
  [OPERATOR_IS_ANY]: {
    key: 'is-any-filter',
    label: (0, _i18n.__)('Is any')
  },
  [OPERATOR_IS_NONE]: {
    key: 'is-none-filter',
    label: (0, _i18n.__)('Is none')
  },
  [OPERATOR_IS_ALL]: {
    key: 'is-all-filter',
    label: (0, _i18n.__)('Is all')
  },
  [OPERATOR_IS_NOT_ALL]: {
    key: 'is-not-all-filter',
    label: (0, _i18n.__)('Is not all')
  }
};
const SORTING_DIRECTIONS = exports.SORTING_DIRECTIONS = ['asc', 'desc'];
const sortArrows = exports.sortArrows = {
  asc: '↑',
  desc: '↓'
};
const sortValues = exports.sortValues = {
  asc: 'ascending',
  desc: 'descending'
};
const sortLabels = exports.sortLabels = {
  asc: (0, _i18n.__)('Sort ascending'),
  desc: (0, _i18n.__)('Sort descending')
};
const sortIcons = exports.sortIcons = {
  asc: _icons.arrowUp,
  desc: _icons.arrowDown
};

// View layouts.
const LAYOUT_TABLE = exports.LAYOUT_TABLE = 'table';
const LAYOUT_GRID = exports.LAYOUT_GRID = 'grid';
const LAYOUT_LIST = exports.LAYOUT_LIST = 'list';
