// @wordpress/edit-post/build/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  initializeEditor: true,
  reinitializeEditor: true,
  __experimentalMainDashboardButton: true,
  __experimentalFullscreenModeClose: true,
  store: true
};
Object.defineProperty(exports, "__experimentalFullscreenModeClose", {
  enumerable: true,
  get: function () {
    return _fullscreenModeClose.default;
  }
});
exports.__experimentalMainDashboardButton = void 0;
exports.initializeEditor = initializeEditor;
exports.reinitializeEditor = reinitializeEditor;
Object.defineProperty(exports, "store", {
  enumerable: true,
  get: function () {
    return _store.store;
  }
});
var _blocks = require("@wordpress/blocks");
var _blockLibrary = require("@wordpress/block-library");
var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _preferences = require("@wordpress/preferences");
var _widgets = require("@wordpress/widgets");
var _editor = require("@wordpress/editor");
var _layout = _interopRequireDefault(require("./components/layout"));
var _lockUnlock = require("./lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
var _fullscreenModeClose = _interopRequireDefault(require("./components/back-button/fullscreen-mode-close"));
var _store = require("./store");
var _deprecated2 = require("./deprecated");
Object.keys(_deprecated2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _deprecated2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _deprecated2[key];
    }
  });
});
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  BackButton: __experimentalMainDashboardButton,
  registerCoreBlockBindingsSources
} = (0, _lockUnlock.unlock)(_editor.privateApis);

/**
 * Initializes and returns an instance of Editor.
 *
 * @param {string}  id           Unique identifier for editor instance.
 * @param {string}  postType     Post type of the post to edit.
 * @param {Object}  postId       ID of the post to edit.
 * @param {?Object} settings     Editor settings object.
 * @param {Object}  initialEdits Programmatic edits to apply initially, to be
 *                               considered as non-user-initiated (bypass for
 *                               unsaved changes prompt).
 */
exports.__experimentalMainDashboardButton = __experimentalMainDashboardButton;
function initializeEditor(id, postType, postId, settings, initialEdits) {
  const isMediumOrBigger = window.matchMedia('(min-width: 782px)').matches;
  const target = document.getElementById(id);
  const root = (0, _element.createRoot)(target);
  (0, _data.dispatch)(_preferences.store).setDefaults('core/edit-post', {
    fullscreenMode: true,
    themeStyles: true,
    welcomeGuide: true,
    welcomeGuideTemplate: true
  });
  (0, _data.dispatch)(_preferences.store).setDefaults('core', {
    allowRightClickOverrides: true,
    editorMode: 'visual',
    editorTool: 'edit',
    fixedToolbar: false,
    hiddenBlockTypes: [],
    inactivePanels: [],
    openPanels: ['post-status'],
    showBlockBreadcrumbs: true,
    showIconLabels: false,
    showListViewByDefault: false,
    enableChoosePatternModal: true,
    isPublishSidebarEnabled: true
  });
  if (window.__experimentalMediaProcessing) {
    (0, _data.dispatch)(_preferences.store).setDefaults('core/media', {
      requireApproval: true,
      optimizeOnUpload: true
    });
  }
  (0, _data.dispatch)(_blocks.store).reapplyBlockTypeFilters();

  // Check if the block list view should be open by default.
  // If `distractionFree` mode is enabled, the block list view should not be open.
  // This behavior is disabled for small viewports.
  if (isMediumOrBigger && (0, _data.select)(_preferences.store).get('core', 'showListViewByDefault') && !(0, _data.select)(_preferences.store).get('core', 'distractionFree')) {
    (0, _data.dispatch)(_editor.store).setIsListViewOpened(true);
  }
  (0, _blockLibrary.registerCoreBlocks)();
  registerCoreBlockBindingsSources();
  (0, _widgets.registerLegacyWidgetBlock)({
    inserter: false
  });
  (0, _widgets.registerWidgetGroupBlock)({
    inserter: false
  });
  if (globalThis.IS_GUTENBERG_PLUGIN) {
    (0, _blockLibrary.__experimentalRegisterExperimentalCoreBlocks)({
      enableFSEBlocks: settings.__unstableEnableFullSiteEditingBlocks
    });
  }

  // Show a console log warning if the browser is not in Standards rendering mode.
  const documentMode = document.compatMode === 'CSS1Compat' ? 'Standards' : 'Quirks';
  if (documentMode !== 'Standards') {
    // eslint-disable-next-line no-console
    console.warn("Your browser is using Quirks Mode. \nThis can cause rendering issues such as blocks overlaying meta boxes in the editor. Quirks Mode can be triggered by PHP errors or HTML code appearing before the opening <!DOCTYPE html>. Try checking the raw page source or your site's PHP error log and resolving errors there, removing any HTML before the doctype, or disabling plugins.");
  }

  // This is a temporary fix for a couple of issues specific to Webkit on iOS.
  // Without this hack the browser scrolls the mobile toolbar off-screen.
  // Once supported in Safari we can replace this in favor of preventScroll.
  // For details see issue #18632 and PR #18686
  // Specifically, we scroll `interface-interface-skeleton__body` to enable a fixed top toolbar.
  // But Mobile Safari forces the `html` element to scroll upwards, hiding the toolbar.

  const isIphone = window.navigator.userAgent.indexOf('iPhone') !== -1;
  if (isIphone) {
    window.addEventListener('scroll', event => {
      const editorScrollContainer = document.getElementsByClassName('interface-interface-skeleton__body')[0];
      if (event.target === document) {
        // Scroll element into view by scrolling the editor container by the same amount
        // that Mobile Safari tried to scroll the html element upwards.
        if (window.scrollY > 100) {
          editorScrollContainer.scrollTop = editorScrollContainer.scrollTop + window.scrollY;
        }
        // Undo unwanted scroll on html element, but only in the visual editor.
        if (document.getElementsByClassName('is-mode-visual')[0]) {
          window.scrollTo(0, 0);
        }
      }
    });
  }

  // Prevent the default browser action for files dropped outside of dropzones.
  window.addEventListener('dragover', e => e.preventDefault(), false);
  window.addEventListener('drop', e => e.preventDefault(), false);
  root.render(/*#__PURE__*/(0, _jsxRuntime.jsx)(_element.StrictMode, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_layout.default, {
      settings: settings,
      postId: postId,
      postType: postType,
      initialEdits: initialEdits
    })
  }));
  return root;
}

/**
 * Used to reinitialize the editor after an error. Now it's a deprecated noop function.
 */
function reinitializeEditor() {
  (0, _deprecated.default)('wp.editPost.reinitializeEditor', {
    since: '6.2',
    version: '6.3'
  });
}
