// @wordpress/fields/build-module/fields/template/template-edit.js
/**
 * WordPress dependencies
 */
import { useCallback, useMemo, useState } from '@wordpress/element';
// @ts-ignore
import { parse } from '@wordpress/blocks';
import { store as coreStore } from '@wordpress/core-data';
/**
 * Internal dependencies
 */
// @ts-expect-error block-editor is not typed correctly.
import { __experimentalBlockPatternsList as BlockPatternsList } from '@wordpress/block-editor';
import { Button, Dropdown, MenuGroup, MenuItem, Modal } from '@wordpress/components';
import { useAsyncList } from '@wordpress/compose';
import { useSelect } from '@wordpress/data';
import { decodeEntities } from '@wordpress/html-entities';
import { __ } from '@wordpress/i18n';
import { getItemTitle } from '../../actions/utils';
import { unlock } from '../../lock-unlock';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const EMPTY_ARRAY = [];
export const TemplateEdit = ({
  data,
  field,
  onChange
}) => {
  const {
    id
  } = field;
  const postType = data.type;
  const postId = typeof data.id === 'number' ? data.id : parseInt(data.id, 10);
  const slug = data.slug;
  const {
    canSwitchTemplate,
    templates
  } = useSelect(select => {
    var _select$getEntityReco;
    const allTemplates = (_select$getEntityReco = select(coreStore).getEntityRecords('postType', 'wp_template', {
      per_page: -1,
      post_type: postType
    })) !== null && _select$getEntityReco !== void 0 ? _select$getEntityReco : EMPTY_ARRAY;
    const {
      getHomePage,
      getPostsPageId
    } = unlock(select(coreStore));
    const isPostsPage = getPostsPageId() === +postId;
    const isFrontPage = postType === 'page' && getHomePage()?.postId === +postId;
    const allowSwitchingTemplate = !isPostsPage && !isFrontPage;
    return {
      templates: allTemplates,
      canSwitchTemplate: allowSwitchingTemplate
    };
  }, [postId, postType]);
  const templatesAsPatterns = useMemo(() => {
    if (!canSwitchTemplate) {
      return [];
    }
    return templates.filter(template => template.is_custom && template.slug !== data.template &&
    // Skip empty templates.
    !!template.content.raw).map(template => ({
      name: template.slug,
      blocks: parse(template.content.raw),
      title: decodeEntities(template.title.rendered),
      id: template.id
    }));
  }, [canSwitchTemplate, data.template, templates]);
  const shownTemplates = useAsyncList(templatesAsPatterns);
  const value = field.getValue({
    item: data
  });
  const foundTemplate = templates.find(template => template.slug === value);
  const currentTemplate = useSelect(select => {
    if (foundTemplate) {
      return foundTemplate;
    }
    let slugToCheck;
    // In `draft` status we might not have a slug available, so we use the `single`
    // post type templates slug(ex page, single-post, single-product etc..).
    // Pages do not need the `single` prefix in the slug to be prioritized
    // through template hierarchy.
    if (slug) {
      slugToCheck = postType === 'page' ? `${postType}-${slug}` : `single-${postType}-${slug}`;
    } else {
      slugToCheck = postType === 'page' ? 'page' : `single-${postType}`;
    }
    if (postType) {
      const templateId = select(coreStore).getDefaultTemplateId({
        slug: slugToCheck
      });
      return select(coreStore).getEntityRecord('postType', 'wp_template', templateId);
    }
  }, [foundTemplate, postType, slug]);
  const [showModal, setShowModal] = useState(false);
  const onChangeControl = useCallback(newValue => onChange({
    [id]: newValue
  }), [id, onChange]);
  return /*#__PURE__*/_jsxs("fieldset", {
    className: "fields-controls__template",
    children: [/*#__PURE__*/_jsx(Dropdown, {
      popoverProps: {
        placement: 'bottom-start'
      },
      renderToggle: ({
        onToggle
      }) => /*#__PURE__*/_jsx(Button, {
        __next40pxDefaultSize: true,
        variant: "tertiary",
        size: "compact",
        onClick: onToggle,
        children: currentTemplate ? getItemTitle(currentTemplate) : ''
      }),
      renderContent: ({
        onToggle
      }) => /*#__PURE__*/_jsxs(MenuGroup, {
        children: [/*#__PURE__*/_jsx(MenuItem, {
          onClick: () => {
            setShowModal(true);
            onToggle();
          },
          children: __('Change template')
        }),
        // The default template in a post is indicated by an empty string
        value !== '' && /*#__PURE__*/_jsx(MenuItem, {
          onClick: () => {
            onChangeControl('');
            onToggle();
          },
          children: __('Use default template')
        })]
      })
    }), showModal && /*#__PURE__*/_jsx(Modal, {
      title: __('Choose a template'),
      onRequestClose: () => setShowModal(false),
      overlayClassName: "fields-controls__template-modal",
      isFullScreen: true,
      children: /*#__PURE__*/_jsx("div", {
        className: "fields-controls__template-content",
        children: /*#__PURE__*/_jsx(BlockPatternsList, {
          label: __('Templates'),
          blockPatterns: templatesAsPatterns,
          shownPatterns: shownTemplates,
          onClickPattern: template => {
            onChangeControl(template.name);
            setShowModal(false);
          }
        })
      })
    })]
  });
};
