// @wordpress/core-data/build/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "__experimentalUseEntityRecord", {
  enumerable: true,
  get: function () {
    return _useEntityRecord.__experimentalUseEntityRecord;
  }
});
Object.defineProperty(exports, "__experimentalUseEntityRecords", {
  enumerable: true,
  get: function () {
    return _useEntityRecords.__experimentalUseEntityRecords;
  }
});
Object.defineProperty(exports, "__experimentalUseResourcePermissions", {
  enumerable: true,
  get: function () {
    return _useResourcePermissions.__experimentalUseResourcePermissions;
  }
});
Object.defineProperty(exports, "useEntityBlockEditor", {
  enumerable: true,
  get: function () {
    return _useEntityBlockEditor.default;
  }
});
Object.defineProperty(exports, "useEntityId", {
  enumerable: true,
  get: function () {
    return _useEntityId.default;
  }
});
Object.defineProperty(exports, "useEntityProp", {
  enumerable: true,
  get: function () {
    return _useEntityProp.default;
  }
});
Object.defineProperty(exports, "useEntityRecord", {
  enumerable: true,
  get: function () {
    return _useEntityRecord.default;
  }
});
Object.defineProperty(exports, "useEntityRecords", {
  enumerable: true,
  get: function () {
    return _useEntityRecords.default;
  }
});
Object.defineProperty(exports, "useResourcePermissions", {
  enumerable: true,
  get: function () {
    return _useResourcePermissions.default;
  }
});
var _useEntityRecord = _interopRequireWildcard(require("./use-entity-record"));
var _useEntityRecords = _interopRequireWildcard(require("./use-entity-records"));
var _useResourcePermissions = _interopRequireWildcard(require("./use-resource-permissions"));
var _useEntityBlockEditor = _interopRequireDefault(require("./use-entity-block-editor"));
var _useEntityId = _interopRequireDefault(require("./use-entity-id"));
var _useEntityProp = _interopRequireDefault(require("./use-entity-prop"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
