// @wordpress/fields/build-module/fields/page-title/view.js
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useSelect } from '@wordpress/data';
import { store as coreStore } from '@wordpress/core-data';
import { privateApis as componentsPrivateApis } from '@wordpress/components';

/**
 * Internal dependencies
 */

import { BaseTitleView } from '../title/view';
import { unlock } from '../../lock-unlock';
import { jsx as _jsx } from "react/jsx-runtime";
const {
  Badge
} = unlock(componentsPrivateApis);
export default function PageTitleView({
  item
}) {
  const {
    frontPageId,
    postsPageId
  } = useSelect(select => {
    const {
      getEntityRecord
    } = select(coreStore);
    const siteSettings = getEntityRecord('root', 'site');
    return {
      frontPageId: siteSettings?.page_on_front,
      postsPageId: siteSettings?.page_for_posts
    };
  }, []);
  return /*#__PURE__*/_jsx(BaseTitleView, {
    item: item,
    className: "fields-field__page-title",
    children: [frontPageId, postsPageId].includes(item.id) && /*#__PURE__*/_jsx(Badge, {
      children: item.id === frontPageId ? __('Homepage') : __('Posts Page')
    })
  });
}
