// @wordpress/editor/build/components/editor-help/move-blocks.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _i18n = require("@wordpress/i18n");
var _style = _interopRequireDefault(require("./style.scss"));
var _viewSections = require("./view-sections");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const MoveBlocks = () => {
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailImage, {
      source: require('./images/drag-and-drop-light.png'),
      sourceDarkMode: require('./images/drag-and-drop-dark.png')
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: _style.default.helpDetailContainer,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailSectionHeadingText, {
        text: (0, _i18n.__)('Drag & drop'),
        badge: (0, _i18n.__)('NEW')
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailBodyText, {
        text: (0, _i18n.__)('Drag & drop makes rearranging blocks a breeze. Press and hold on a block, then drag it to its new location and release.')
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailImage, {
      source: require('./images/move-light.png'),
      sourceDarkMode: require('./images/move-dark.png')
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: _style.default.helpDetailContainer,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailSectionHeadingText, {
        text: (0, _i18n.__)('Arrow buttons')
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSections.HelpDetailBodyText, {
        text: (0, _i18n.__)('You can also rearrange blocks by tapping a block and then tapping the up and down arrows that appear on the bottom left side of the block to move it up or down.')
      })]
    })]
  });
};
var _default = exports.default = MoveBlocks;
