// @wordpress/editor/build/components/editor-help/view-sections.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HelpDetailSectionHeadingText = exports.HelpDetailImage = exports.HelpDetailBodyText = void 0;
var _reactNative = require("react-native");
var _compose = require("@wordpress/compose");
var _style = _interopRequireDefault(require("./style.scss"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const HelpDetailBodyText = ({
  text
}) => {
  const bodyStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailBody, _style.default.helpDetailBodyDark);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
    selectable: true,
    style: bodyStyle,
    children: text
  });
};
exports.HelpDetailBodyText = HelpDetailBodyText;
const HelpDetailSectionHeadingText = ({
  text,
  badge
}) => {
  const headingTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailSectionHeadingText, _style.default.helpDetailSectionHeadingTextDark);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
    style: _style.default.helpDetailSectionHeading,
    children: [badge && /*#__PURE__*/(0, _jsxRuntime.jsx)(HelpDetailBadge, {
      text: badge
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
      accessibilityRole: "header",
      selectable: true,
      style: headingTextStyle,
      children: text
    })]
  });
};
exports.HelpDetailSectionHeadingText = HelpDetailSectionHeadingText;
const HelpDetailImage = ({
  accessible,
  accessibilityLabel,
  source,
  sourceDarkMode
}) => {
  const imageStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailImage, _style.default.helpDetailImageDark);
  const darkModeEnabled = (0, _compose.usePreferredColorScheme)() === 'dark';
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Image, {
    accessible: accessible,
    accessibilityLabel: accessibilityLabel,
    source: darkModeEnabled && sourceDarkMode ? sourceDarkMode : source,
    style: imageStyle
  });
};
exports.HelpDetailImage = HelpDetailImage;
const HelpDetailBadge = ({
  text
}) => {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: _style.default.helpDetailBadgeContainer,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
      style: _style.default.helpDetailBadgeText,
      children: text
    })
  });
};
