// @wordpress/dataviews/build-module/dataviews-layouts/table/column-primary.js
/**
 * WordPress dependencies
 */
import { __experimentalHStack as HStack, __experimentalVStack as VStack } from '@wordpress/components';

/**
 * Internal dependencies
 */

import getClickableItemProps from '../utils/get-clickable-item-props';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
function ColumnPrimary({
  item,
  level,
  titleField,
  mediaField,
  descriptionField,
  onClickItem,
  isItemClickable
}) {
  const clickableProps = getClickableItemProps({
    item,
    isItemClickable,
    onClickItem,
    className: 'dataviews-view-table__cell-content-wrapper dataviews-title-field'
  });
  return /*#__PURE__*/_jsxs(HStack, {
    spacing: 3,
    justify: "flex-start",
    children: [mediaField && /*#__PURE__*/_jsx("div", {
      className: "dataviews-view-table__cell-content-wrapper dataviews-column-primary__media",
      children: /*#__PURE__*/_jsx(mediaField.render, {
        item: item
      })
    }), /*#__PURE__*/_jsxs(VStack, {
      spacing: 0,
      children: [titleField && /*#__PURE__*/_jsxs("div", {
        ...clickableProps,
        children: [level !== undefined && /*#__PURE__*/_jsxs("span", {
          className: "dataviews-view-table__level",
          children: ['—'.repeat(level), "\xA0"]
        }), /*#__PURE__*/_jsx(titleField.render, {
          item: item
        })]
      }), descriptionField && /*#__PURE__*/_jsx(descriptionField.render, {
        item: item
      })]
    })]
  });
}
export default ColumnPrimary;
