// @wordpress/dataviews/build-module/dataviews-layouts/table/density-picker.js
/**
 * WordPress dependencies
 */
import { __experimentalToggleGroupControl as ToggleGroupControl, __experimentalToggleGroupControlOption as ToggleGroupControlOption } from '@wordpress/components';
import { __, _x } from '@wordpress/i18n';
import { useContext } from '@wordpress/element';

/**
 * Internal dependencies
 */
import DataViewsContext from '../../components/dataviews-context';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export default function DensityPicker() {
  const context = useContext(DataViewsContext);
  const view = context.view;
  return /*#__PURE__*/_jsxs(ToggleGroupControl, {
    __nextHasNoMarginBottom: true,
    size: "__unstable-large",
    label: __('Density'),
    value: view.layout?.density || 'balanced',
    onChange: value => {
      context.onChangeView({
        ...view,
        layout: {
          ...view.layout,
          density: value
        }
      });
    },
    isBlock: true,
    children: [/*#__PURE__*/_jsx(ToggleGroupControlOption, {
      value: "comfortable",
      label: _x('Comfortable', 'Density option for DataView layout')
    }, "comfortable"), /*#__PURE__*/_jsx(ToggleGroupControlOption, {
      value: "balanced",
      label: _x('Balanced', 'Density option for DataView layout')
    }, "balanced"), /*#__PURE__*/_jsx(ToggleGroupControlOption, {
      value: "compact",
      label: _x('Compact', 'Density option for DataView layout')
    }, "compact")]
  });
}
