// @wordpress/fields/build/fields/page-title/view.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PageTitleView;
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _components = require("@wordpress/components");
var _view = require("../title/view");
var _lockUnlock = require("../../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Badge
} = (0, _lockUnlock.unlock)(_components.privateApis);
function PageTitleView({
  item
}) {
  const {
    frontPageId,
    postsPageId
  } = (0, _data.useSelect)(select => {
    const {
      getEntityRecord
    } = select(_coreData.store);
    const siteSettings = getEntityRecord('root', 'site');
    return {
      frontPageId: siteSettings?.page_on_front,
      postsPageId: siteSettings?.page_for_posts
    };
  }, []);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_view.BaseTitleView, {
    item: item,
    className: "fields-field__page-title",
    children: [frontPageId, postsPageId].includes(item.id) && /*#__PURE__*/(0, _jsxRuntime.jsx)(Badge, {
      children: item.id === frontPageId ? (0, _i18n.__)('Homepage') : (0, _i18n.__)('Posts Page')
    })
  });
}
