// @wordpress/edit-post/build/components/more-menu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _editor = require("@wordpress/editor");
var _keycodes = require("@wordpress/keycodes");
var _preferences = require("@wordpress/preferences");
var _lockUnlock = require("../../lock-unlock");
var _managePatternsMenuItem = _interopRequireDefault(require("./manage-patterns-menu-item"));
var _welcomeGuideMenuItem = _interopRequireDefault(require("./welcome-guide-menu-item"));
var _preferencesModal = _interopRequireDefault(require("../preferences-modal"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  ToolsMoreMenuGroup,
  ViewMoreMenuGroup
} = (0, _lockUnlock.unlock)(_editor.privateApis);
const MoreMenu = () => {
  const isLargeViewport = (0, _compose.useViewportMatch)('large');
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [isLargeViewport && /*#__PURE__*/(0, _jsxRuntime.jsx)(ViewMoreMenuGroup, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_preferences.PreferenceToggleMenuItem, {
        scope: "core/edit-post",
        name: "fullscreenMode",
        label: (0, _i18n.__)('Fullscreen mode'),
        info: (0, _i18n.__)('Show and hide the admin user interface'),
        messageActivated: (0, _i18n.__)('Fullscreen mode activated.'),
        messageDeactivated: (0, _i18n.__)('Fullscreen mode deactivated.'),
        shortcut: _keycodes.displayShortcut.secondary('f')
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(ToolsMoreMenuGroup, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_managePatternsMenuItem.default, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_welcomeGuideMenuItem.default, {})]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_preferencesModal.default, {})]
  });
};
var _default = exports.default = MoreMenu;
