// @wordpress/fields/build/fields/status/status-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _statusElements = _interopRequireDefault(require("./status-elements"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function StatusView({
  item
}) {
  const status = _statusElements.default.find(({
    value
  }) => value === item.status);
  const label = status?.label || item.status;
  const icon = status?.icon;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    alignment: "left",
    spacing: 0,
    children: [icon && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "edit-site-post-list__status-icon",
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Icon, {
        icon: icon
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      children: label
    })]
  });
}
var _default = exports.default = StatusView;
