// @wordpress/block-library/build/html/preview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HTMLEditPreview;
var _element = require("@wordpress/element");
var _blockEditor = require("@wordpress/block-editor");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

// Default styles used to unset some of the styles
// that might be inherited from the editor style.
const DEFAULT_STYLES = `
	html,body,:root {
		margin: 0 !important;
		padding: 0 !important;
		overflow: visible !important;
		min-height: auto !important;
	}
`;
function HTMLEditPreview({
  content,
  isSelected
}) {
  const settingStyles = (0, _data.useSelect)(select => select(_blockEditor.store).getSettings().styles, []);
  const styles = (0, _element.useMemo)(() => [DEFAULT_STYLES, ...(0, _blockEditor.transformStyles)((settingStyles !== null && settingStyles !== void 0 ? settingStyles : []).filter(style => style.css))], [settingStyles]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_components.SandBox, {
      html: content,
      styles: styles,
      title: (0, _i18n.__)('Custom HTML Preview'),
      tabIndex: -1
    }), !isSelected && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "block-library-html__preview-overlay"
    })]
  });
}
