// @wordpress/patterns/build/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PatternsMenuItems;
var _blockEditor = require("@wordpress/block-editor");
var _patternConvertButton = _interopRequireDefault(require("./pattern-convert-button"));
var _patternsManageButton = _interopRequireDefault(require("./patterns-manage-button"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternsMenuItems({
  rootClientId
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.BlockSettingsMenuControls, {
    children: ({
      selectedClientIds,
      onClose
    }) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_patternConvertButton.default, {
        clientIds: selectedClientIds,
        rootClientId: rootClientId,
        closeBlockSettingsMenu: onClose
      }), selectedClientIds.length === 1 && /*#__PURE__*/(0, _jsxRuntime.jsx)(_patternsManageButton.default, {
        clientId: selectedClientIds[0]
      })]
    })
  });
}
