// @wordpress/dataviews/build/components/dataviews-footer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DataViewsFooter;
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _dataviewsContext = _interopRequireDefault(require("../dataviews-context"));
var _dataviewsPagination = _interopRequireDefault(require("../dataviews-pagination"));
var _dataviewsBulkActions = require("../dataviews-bulk-actions");
var _constants = require("../../constants");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const EMPTY_ARRAY = [];
function DataViewsFooter() {
  const {
    view,
    paginationInfo: {
      totalItems = 0,
      totalPages
    },
    data,
    actions = EMPTY_ARRAY
  } = (0, _element.useContext)(_dataviewsContext.default);
  const hasBulkActions = (0, _dataviewsBulkActions.useSomeItemHasAPossibleBulkAction)(actions, data) && [_constants.LAYOUT_TABLE, _constants.LAYOUT_GRID].includes(view.type);
  if (!totalItems || !totalPages || totalPages <= 1 && !hasBulkActions) {
    return null;
  }
  return !!totalItems && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalHStack, {
    expanded: false,
    justify: "end",
    className: "dataviews-footer",
    children: [hasBulkActions && /*#__PURE__*/(0, _jsxRuntime.jsx)(_dataviewsBulkActions.BulkActionsFooter, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_dataviewsPagination.default, {})]
  });
}
