// @wordpress/fields/build/actions/duplicate-pattern.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _i18n = require("@wordpress/i18n");
var _patterns = require("@wordpress/patterns");
var _lockUnlock = require("../lock-unlock");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

// @ts-ignore

/**
 * Internal dependencies
 */

// Patterns.
const {
  CreatePatternModalContents,
  useDuplicatePatternProps
} = (0, _lockUnlock.unlock)(_patterns.privateApis);
const duplicatePattern = {
  id: 'duplicate-pattern',
  label: (0, _i18n._x)('Duplicate', 'action label'),
  isEligible: item => item.type !== 'wp_template_part',
  modalHeader: (0, _i18n._x)('Duplicate pattern', 'action label'),
  modalFocusOnMount: 'firstContentElement',
  RenderModal: ({
    items,
    closeModal
  }) => {
    const [item] = items;
    const duplicatedProps = useDuplicatePatternProps({
      pattern: item,
      onSuccess: () => closeModal?.()
    });
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(CreatePatternModalContents, {
      onClose: closeModal,
      confirmLabel: (0, _i18n._x)('Duplicate', 'action label'),
      ...duplicatedProps
    });
  }
};

/**
 * Duplicate action for Pattern.
 */
var _default = exports.default = duplicatePattern;
