// @wordpress/fields/build/actions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "deletePost", {
  enumerable: true,
  get: function () {
    return _deletePost.default;
  }
});
Object.defineProperty(exports, "duplicatePattern", {
  enumerable: true,
  get: function () {
    return _duplicatePattern.default;
  }
});
Object.defineProperty(exports, "duplicatePost", {
  enumerable: true,
  get: function () {
    return _duplicatePost.default;
  }
});
Object.defineProperty(exports, "duplicateTemplatePart", {
  enumerable: true,
  get: function () {
    return _duplicateTemplatePart.default;
  }
});
Object.defineProperty(exports, "exportPattern", {
  enumerable: true,
  get: function () {
    return _exportPattern.default;
  }
});
Object.defineProperty(exports, "permanentlyDeletePost", {
  enumerable: true,
  get: function () {
    return _permanentlyDeletePost.default;
  }
});
Object.defineProperty(exports, "renamePost", {
  enumerable: true,
  get: function () {
    return _renamePost.default;
  }
});
Object.defineProperty(exports, "reorderPage", {
  enumerable: true,
  get: function () {
    return _reorderPage.default;
  }
});
Object.defineProperty(exports, "resetPost", {
  enumerable: true,
  get: function () {
    return _resetPost.default;
  }
});
Object.defineProperty(exports, "restorePost", {
  enumerable: true,
  get: function () {
    return _restorePost.default;
  }
});
Object.defineProperty(exports, "trashPost", {
  enumerable: true,
  get: function () {
    return _trashPost.default;
  }
});
Object.defineProperty(exports, "viewPost", {
  enumerable: true,
  get: function () {
    return _viewPost.default;
  }
});
Object.defineProperty(exports, "viewPostRevisions", {
  enumerable: true,
  get: function () {
    return _viewPostRevisions.default;
  }
});
var _viewPost = _interopRequireDefault(require("./view-post"));
var _reorderPage = _interopRequireDefault(require("./reorder-page"));
var _duplicatePost = _interopRequireDefault(require("./duplicate-post"));
var _renamePost = _interopRequireDefault(require("./rename-post"));
var _resetPost = _interopRequireDefault(require("./reset-post"));
var _duplicatePattern = _interopRequireDefault(require("./duplicate-pattern"));
var _exportPattern = _interopRequireDefault(require("./export-pattern"));
var _viewPostRevisions = _interopRequireDefault(require("./view-post-revisions"));
var _permanentlyDeletePost = _interopRequireDefault(require("./permanently-delete-post"));
var _restorePost = _interopRequireDefault(require("./restore-post"));
var _trashPost = _interopRequireDefault(require("./trash-post"));
var _deletePost = _interopRequireDefault(require("./delete-post"));
var _duplicateTemplatePart = _interopRequireDefault(require("./duplicate-template-part"));
