// @wordpress/editor/build/components/collab-sidebar/comment-author-info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _date = require("@wordpress/date");
var _coreData = require("@wordpress/core-data");
var _data = require("@wordpress/data");
var _blockEditor = require("@wordpress/block-editor");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Render author information for a comment.
 *
 * @param {Object} props        - Component properties.
 * @param {string} props.avatar - URL of the author's avatar.
 * @param {string} props.name   - Name of the author.
 * @param {string} props.date   - Date of the comment.
 *
 * @return {React.ReactNode} The JSX element representing the author's information.
 */function CommentAuthorInfo({
  avatar,
  name,
  date
}) {
  const dateSettings = (0, _date.getSettings)();
  const [dateTimeFormat = dateSettings.formats.time] = (0, _coreData.useEntityProp)('root', 'site', 'time_format');
  const {
    currentUserAvatar,
    currentUserName
  } = (0, _data.useSelect)(select => {
    var _userData$avatar_urls;
    const userData = select(_coreData.store).getCurrentUser();
    const {
      getSettings
    } = select(_blockEditor.store);
    const {
      __experimentalDiscussionSettings
    } = getSettings();
    const defaultAvatar = __experimentalDiscussionSettings?.avatarURL;
    return {
      currentUserAvatar: (_userData$avatar_urls = userData?.avatar_urls[48]) !== null && _userData$avatar_urls !== void 0 ? _userData$avatar_urls : defaultAvatar,
      currentUserName: userData?.name
    };
  }, []);
  const currentDate = new Date();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
      src: avatar !== null && avatar !== void 0 ? avatar : currentUserAvatar,
      className: "editor-collab-sidebar-panel__user-avatar"
      // translators: alt text for user avatar image
      ,
      alt: (0, _i18n.__)('User avatar'),
      width: 32,
      height: 32
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_components.__experimentalVStack, {
      spacing: "0",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        className: "editor-collab-sidebar-panel__user-name",
        children: name !== null && name !== void 0 ? name : currentUserName
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("time", {
        dateTime: (0, _date.dateI18n)('c', date !== null && date !== void 0 ? date : currentDate),
        className: "editor-collab-sidebar-panel__user-time",
        children: (0, _date.dateI18n)(dateTimeFormat, date !== null && date !== void 0 ? date : currentDate)
      })]
    })]
  });
}
var _default = exports.default = CommentAuthorInfo;
