// @wordpress/reusable-blocks/build/components/reusable-blocks-menu-items/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReusableBlocksMenuItems;
var _blockEditor = require("@wordpress/block-editor");
var _reusableBlockConvertButton = _interopRequireDefault(require("./reusable-block-convert-button"));
var _reusableBlocksManageButton = _interopRequireDefault(require("./reusable-blocks-manage-button"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ReusableBlocksMenuItems({
  rootClientId
}) {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.BlockSettingsMenuControls, {
    children: ({
      onClose,
      selectedClientIds
    }) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_reusableBlockConvertButton.default, {
        clientIds: selectedClientIds,
        rootClientId: rootClientId,
        onClose: onClose
      }), selectedClientIds.length === 1 && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reusableBlocksManageButton.default, {
        clientId: selectedClientIds[0]
      })]
    })
  });
}
