// @wordpress/edit-post/build/components/preferences-modal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EditPostPreferencesModal;
var _i18n = require("@wordpress/i18n");
var _preferences = require("@wordpress/preferences");
var _editor = require("@wordpress/editor");
var _lockUnlock = require("../../lock-unlock");
var _metaBoxesSection = _interopRequireDefault(require("./meta-boxes-section"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  PreferenceToggleControl
} = (0, _lockUnlock.unlock)(_preferences.privateApis);
const {
  PreferencesModal
} = (0, _lockUnlock.unlock)(_editor.privateApis);
function EditPostPreferencesModal() {
  const extraSections = {
    general: /*#__PURE__*/(0, _jsxRuntime.jsx)(_metaBoxesSection.default, {
      title: (0, _i18n.__)('Advanced')
    }),
    appearance: /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferenceToggleControl, {
      scope: "core/edit-post",
      featureName: "themeStyles",
      help: (0, _i18n.__)('Make the editor look like your theme.'),
      label: (0, _i18n.__)('Use theme styles')
    })
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesModal, {
    extraSections: extraSections
  });
}
