// @wordpress/fields/build-module/actions/duplicate-pattern.js
/**
 * WordPress dependencies
 */
import { _x } from '@wordpress/i18n';
// @ts-ignore
import { privateApis as patternsPrivateApis } from '@wordpress/patterns';
/**
 * Internal dependencies
 */
import { unlock } from '../lock-unlock';
import { jsx as _jsx } from "react/jsx-runtime";
// Patterns.
const {
  CreatePatternModalContents,
  useDuplicatePatternProps
} = unlock(patternsPrivateApis);
const duplicatePattern = {
  id: 'duplicate-pattern',
  label: _x('Duplicate', 'action label'),
  isEligible: item => item.type !== 'wp_template_part',
  modalHeader: _x('Duplicate pattern', 'action label'),
  modalFocusOnMount: 'firstContentElement',
  RenderModal: ({
    items,
    closeModal
  }) => {
    const [item] = items;
    const duplicatedProps = useDuplicatePatternProps({
      pattern: item,
      onSuccess: () => closeModal?.()
    });
    return /*#__PURE__*/_jsx(CreatePatternModalContents, {
      onClose: closeModal,
      confirmLabel: _x('Duplicate', 'action label'),
      ...duplicatedProps
    });
  }
};

/**
 * Duplicate action for Pattern.
 */
export default duplicatePattern;
