// @wordpress/fields/build-module/actions/duplicate-template-part.js
/**
 * WordPress dependencies
 */
import { useDispatch } from '@wordpress/data';
import { _x, sprintf } from '@wordpress/i18n';
import { store as noticesStore } from '@wordpress/notices';
import { useMemo } from '@wordpress/element';
// @ts-ignore
import { parse } from '@wordpress/blocks';

/**
 * Internal dependencies
 */

import { CreateTemplatePartModalContents } from '../components/create-template-part-modal';
import { getItemTitle } from './utils';

/**
 * This action is used to duplicate a template part.
 */
import { jsx as _jsx } from "react/jsx-runtime";
const duplicateTemplatePart = {
  id: 'duplicate-template-part',
  label: _x('Duplicate', 'action label'),
  isEligible: item => item.type === 'wp_template_part',
  modalHeader: _x('Duplicate template part', 'action label'),
  modalFocusOnMount: 'firstContentElement',
  RenderModal: ({
    items,
    closeModal
  }) => {
    const [item] = items;
    const blocks = useMemo(() => {
      var _item$blocks;
      return (_item$blocks = item.blocks) !== null && _item$blocks !== void 0 ? _item$blocks : parse(typeof item.content === 'string' ? item.content : item.content.raw, {
        __unstableSkipMigrationLogs: true
      });
    }, [item.content, item.blocks]);
    const {
      createSuccessNotice
    } = useDispatch(noticesStore);
    function onTemplatePartSuccess(templatePart) {
      createSuccessNotice(sprintf(
      // translators: %s: The new template part's title e.g. 'Call to action (copy)'.
      _x('"%s" duplicated.', 'template part'), getItemTitle(templatePart)), {
        type: 'snackbar',
        id: 'edit-site-patterns-success'
      });
      closeModal?.();
    }
    return /*#__PURE__*/_jsx(CreateTemplatePartModalContents, {
      blocks: blocks,
      defaultArea: item.area,
      defaultTitle: sprintf(/* translators: %s: Existing template part title */
      _x('%s (Copy)', 'template part'), getItemTitle(item)),
      onCreate: onTemplatePartSuccess,
      onError: closeModal,
      confirmLabel: _x('Duplicate', 'action label'),
      closeModal: closeModal !== null && closeModal !== void 0 ? closeModal : () => {}
    });
  }
};
/**
 * Duplicate action for TemplatePart.
 */
export default duplicateTemplatePart;
