// @wordpress/block-library/build/preformatted/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreformattedEdit;
var _i18n = require("@wordpress/i18n");
var _blockEditor = require("@wordpress/block-editor");
var _blocks = require("@wordpress/blocks");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

function PreformattedEdit({
  attributes,
  mergeBlocks,
  setAttributes,
  onRemove,
  insertBlocksAfter,
  style
}) {
  const {
    content
  } = attributes;
  const blockProps = (0, _blockEditor.useBlockProps)({
    style
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_blockEditor.RichText, {
    tagName: "pre",
    identifier: "content",
    preserveWhiteSpace: true,
    value: content,
    onChange: nextContent => {
      setAttributes({
        content: nextContent
      });
    },
    onRemove: onRemove,
    "aria-label": (0, _i18n.__)('Preformatted text'),
    placeholder: (0, _i18n.__)('Write preformatted text…'),
    onMerge: mergeBlocks,
    ...blockProps,
    __unstablePastePlainText: true,
    __unstableOnSplitAtDoubleLineEnd: () => insertBlocksAfter((0, _blocks.createBlock)((0, _blocks.getDefaultBlockName)()))
  });
}
